{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {
    "toc": true
   },
   "source": [
    "<h1>Data parsing<span class=\"tocSkip\"></span></h1>\n",
    "<div class=\"toc\"><ul class=\"toc-item\"><li><span><a href=\"#Import-dependencies\" data-toc-modified-id=\"Import-dependencies-1\">Import dependencies</a></span></li><li><span><a href=\"#Load-data\" data-toc-modified-id=\"Load-data-2\">Load data</a></span></li><li><span><a href=\"#Data-exploration\" data-toc-modified-id=\"Data-exploration-3\">Data exploration</a></span><ul class=\"toc-item\"><li><span><a href=\"#Get-a-list-of-standalone-paragraphs-from-each-article's-body\" data-toc-modified-id=\"Get-a-list-of-standalone-paragraphs-from-each-article's-body-3.1\">Get a list of standalone paragraphs from each article's body</a></span></li><li><span><a href=\"#Get-a-list-of-sections-(in-each-section-are-one-or-more-paragraphs)--from-each-article's-body\" data-toc-modified-id=\"Get-a-list-of-sections-(in-each-section-are-one-or-more-paragraphs)--from-each-article's-body-3.2\">Get a list of sections (in each section are one or more paragraphs)  from each article's body</a></span></li><li><span><a href=\"#Get-article's-body\" data-toc-modified-id=\"Get-article's-body-3.3\">Get article's body</a></span></li><li><span><a href=\"#Get-article's-abstract\" data-toc-modified-id=\"Get-article's-abstract-3.4\">Get article's abstract</a></span></li><li><span><a href=\"#Get-article's-categories\" data-toc-modified-id=\"Get-article's-categories-3.5\">Get article's categories</a></span></li><li><span><a href=\"#Get-article's-title\" data-toc-modified-id=\"Get-article's-title-3.6\">Get article's title</a></span></li><li><span><a href=\"#Get-final-dataframe\" data-toc-modified-id=\"Get-final-dataframe-3.7\">Get final dataframe</a></span></li><li><span><a href=\"#Get-number-of-pages\" data-toc-modified-id=\"Get-number-of-pages-3.8\">Get number of pages</a></span></li><li><span><a href=\"#Get-authors-list\" data-toc-modified-id=\"Get-authors-list-3.9\">Get authors list</a></span></li><li><span><a href=\"#Get-author's-affiliation\" data-toc-modified-id=\"Get-author's-affiliation-3.10\">Get author's affiliation</a></span></li><li><span><a href=\"#Get-number-of-figures\" data-toc-modified-id=\"Get-number-of-figures-3.11\">Get number of figures</a></span></li></ul></li></ul></div>"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Import dependencies"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "import os, shutil\n",
    "import warnings\n",
    "import random\n",
    "warnings.simplefilter(\"ignore\")\n",
    "\n",
    "from shutil import copyfile\n",
    "from distutils.dir_util import copy_tree"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Load data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "input_dir = \"../datasets/2016_testing_df/\"\n",
    "\n",
    "df = sqlContext.read.format('com.databricks.spark.xml').option(\"rowTag\", \"record\")\\\n",
    "    .load(input_dir)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "114"
      ]
     },
     "execution_count": 3,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df.rdd.count()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Data exploration"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Get a list of standalone paragraphs from each article's body"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "The paragraphs can be found: in **body/p** or simply in **p** tag from the root. But the results are the same."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "[(0,\n",
       "  ['The mature 26S proteasome consists of at least 33 distinct subunits. Fourteen of them (α1-7 and β1-7) form the 20S core particle (CP), a barrel-shaped structure that encloses three types of peptidase activities (trypsin-like, caspase-like and chymotrypsin-like). The remaining 19 subunits (Rpt1-6, Rpn1-3, 5-13 and 15) constitute the 19S regulatory particle (RP) that caps the CP on one or both ends. Protein substrates destined for proteasomal degradation are captured and processed by the 19S RP before they are threaded into the 20S CP for proteolysis. During this process, the ATPase subunits (Rpt1-6) play key roles in substrate engagement, unfolding, translocation and CP gate opening']),\n",
       " (1,\n",
       "  ['Purely computational approaches have been used to perform virtual screens across multiple mechanisms of action']),\n",
       " (2,\n",
       "  ['The zebrafish has recently emerged as a useful vertebrate model system to study neural circuits and behavior, but tools to modulate neurons in freely behaving animals are limited. As a poikilotherm that lives in water, zebrafish are amenable to thermal and pharmacological perturbations. We exploit these properties by using transient receptor potential (TRP) channels to inducibly activate and ablate specific larval zebrafish (']),\n",
       " (3,\n",
       "  ['Somatic copy number alterations (SCNAs), including chromosome arm-level copy changes as well as focal amplifications and deletions, are central events in cancer pathogenesis']),\n",
       " (4,\n",
       "  ['Interrupting the flow of information by disconnecting brain regions from one another has been, and remains, a powerful means to investigate interactions between putative components of neural circuits. Previous work from this laboratory has demonstrated that functionally disconnecting rhinal cortex (Rh) and orbitofrontal cortex (OFC) leads to altered estimates of expected value through reward-stimulus association']),\n",
       " (5,\n",
       "  ['Genotype-phenotype correlations have been reported for these CdLS-associated mutations. For example, ']),\n",
       " (6,\n",
       "  ['The transition of the uterus from a quiescent to contractile organ results from a complex interplay between inflammatory and hormonal signals yet the relative contribution of each to the contractile phenotype remains poorly understood. In the mouse, parturition at term appears to be initiated by fetal lung maturation, which drives steroid receptor cofactor-dependent secretion of surfactant proteins into the amniotic fluid, migration of activated macrophages into the uterus, and induction of the inflammatory transcription factor nuclear factor kappa light chain enhancer of activated B cells (NFkB). This leads to the upregulation of inflammatory and contraction-associated genes within the uterus, including those encoding for the principal prostaglandin synthase enzyme, cyclooxygenase-2 (COX-2). Increased levels of prostaglandin F2 alpha (PGF']),\n",
       " (7,\n",
       "  ['Among these cutaneous manifestations, the blueberry muffin lesions are described as a rare cutaneous eruption, presenting with transient, red-violaceous papules, attributable to a marked dermal hematopoietic activity classically explained by congenital infections and hematological diseases. An unusual presentation of blueberry muffin lesions associated with marked elevation of the circulating nucleated red blood cells (NRBC) and with ultrasound findings of brain ischemic lesions is described in a neonate born from a mother affected by severe anemia during pregnancy.']),\n",
       " (8,\n",
       "  ['A small number of divergent PRD class homeobox genes have very restricted phylogenetic distribution and offer an opportunity to examine the evolutionary origin and functional recruitment of novel homeobox genes in mammals. ']),\n",
       " (9,\n",
       "  ['At birth, lung aeration triggers a functional re-organisation of the infant’s circulation, largely by stimulating an increase in pulmonary blood flow (PBF) ['])]"
      ]
     },
     "execution_count": 4,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "def parse_paragraphs_list(list_of_paragraphs):\n",
    "    list_of_paragraphs_to_ret = []\n",
    "    if list_of_paragraphs != None:\n",
    "        for paragraph in list_of_paragraphs:\n",
    "            if paragraph != None:\n",
    "                if type(paragraph) == str:\n",
    "                    list_of_paragraphs_to_ret.append(paragraph)\n",
    "                else:\n",
    "                    if paragraph._VALUE != None:\n",
    "                        list_of_paragraphs_to_ret.append(paragraph._VALUE)\n",
    "    return list_of_paragraphs_to_ret\n",
    "\n",
    "def get_paragraphs_list_from_body(input_dir):\n",
    "    df =   spark.read \\\n",
    "                .format('com.databricks.spark.xml') \\\n",
    "                .options(rowTag='record')\\\n",
    "                .options(rowTag='body')\\\n",
    "                .load(input_dir)\n",
    "    return  df.select(\"p\")\\\n",
    "              .rdd\\\n",
    "              .map(lambda row: parse_paragraphs_list(row['p']))\\\n",
    "              .zipWithIndex()\\\n",
    "              .map(lambda record: (record[1], record[0]))\\\n",
    "\n",
    "\n",
    "get_paragraphs_list_from_body(input_dir).take(10)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Get a list of sections (in each section are one or more paragraphs)  from each article's body"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "The sections are the same as in **metadata/article/body**"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {
    "scrolled": false
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "[(0,\n",
       "  [['The 26S proteasome is an essential protein complex responsible for degrading the majority of cellular proteins in eukaryotes']]),\n",
       " (1,\n",
       "  [['Advances in molecular biology have led to an unprecedented ability to profile the genetic- and pathway-level changes that occur in disease']]),\n",
       " (2,\n",
       "  [['We fused rat TRPV1 (NM_031982), containing the E600K mutation that increases sensitivity to Csn by over 10–fold ']]),\n",
       " (3,\n",
       "  [['BICR-31 was obtained from Sigma-Aldrich, and A549, NCI-H2009, NCI-H358, HCC95, and Ishikawa cells were obtained from the American Type Culture Collection. Cells were cultured in RPMI 1640 medium supplemented with 10% FBS and 1% penicillin-streptomycin.'],\n",
       "   ['Chromatin-immunoprecipitation followed by massive parallel sequencing (ChIP-seq) was performed as previously described']]),\n",
       " (4,\n",
       "  [['Lentivirus was produced at titers >10E9 using methods similar to those described by Han et al. ']]),\n",
       " (5,\n",
       "  [['Cornelia de Lange syndrome (CdLS, OMIM 608667) is the prototype for cohesinopathy disorders, which are developmental disorders with mutations in an evolutionarily conserved complex that functions in sister chromatin cohesion. However, the complex is also implicated in an increasing number of functions, including transcription regulation, DNA repair, chromosome condensation, and homolog pairing [']]),\n",
       " (6,\n",
       "  [['Preterm labor (PTL), defined as labor onset before 37\\xa0weeks of gestation, is the leading cause of neonatal mortality in children under 5\\xa0years of age, with incidence rates continuing to rise in both developing and developed countries [']]),\n",
       " (7,\n",
       "  [['Neonatal cutaneous eruption differential diagnosis is challenging because multiple conditions may play a causative role in the pathogenesis of a cutaneous rash in a neonate such as congenital infections, hemolytic disease, twin-twin transfusion syndrome, fetomaternal hemorrhage, intracranial bleeding, and neoplastic, autoimmune or hematological disorders [']]),\n",
       " (8,\n",
       "  [['Homeobox genes encode transcription factors with a recognisable DNA-binding domain, the homeodomain, and most play regulatory roles in cell fate determination and embryonic patterning in animals. The majority of metazoan homeobox genes, including those of human, are members of small gene families that have been highly conserved through the evolutionary diversification of the bilaterians [']]),\n",
       " (9,\n",
       "  [['The transition from fetal to newborn life represents one of the greatest physiological challenges that any human will encounter. Once the umbilical cord is clamped, infants must clear their airways of liquid to allow the onset of pulmonary gas exchange and the cardiovascular system must undergo a major structural and functional re-organisation ['],\n",
       "   ['It has long been recognized that UCC is not just a symbolic separation of the infant from the mother, but can have a major impact on the infant’s well-being after birth. Indeed, the argument of when the umbilical cord should be clamped extends back centuries, at least to Aristotle in 300 BC. In 1801, Erasmus Darwin suggested that, ']])]"
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "def parse_sections_list(list_of_sections):\n",
    "    if list_of_sections != None:\n",
    "        list_of_sections = [parse_paragraphs_list(section.p) for section in list_of_sections if section.p != None]\n",
    "    else:\n",
    "        list_of_sections = []\n",
    "    return list_of_sections\n",
    "\n",
    "def get_sections_list(input_dir):\n",
    "    df =   spark.read \\\n",
    "                .format('com.databricks.spark.xml') \\\n",
    "                .options(rowTag='record')\\\n",
    "                .options(rowTag='body')\\\n",
    "                .load(input_dir)\n",
    "    return  df.select(\"sec\")\\\n",
    "              .rdd\\\n",
    "              .map(lambda row: parse_sections_list(row['sec']), )\\\n",
    "              .zipWithIndex()\\\n",
    "              .map(lambda record: (record[1], record[0]))\\\n",
    "\n",
    "get_sections_list(input_dir).take(10)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Get article's body"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "The body of the article is going to be build in the following manner:\n",
    "-  we will concatenate each standalone paragraph (1)\n",
    "-  we will concatenate each paragraph from a section\n",
    "-  we will concatenate each section (3)\n",
    "-  finally, we will concatenate the the results from (1) and (3)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [],
   "source": [
    "def build_body(row_sec, row_p):\n",
    "    concat_parsed_sections = ' '.join([' '.join(section) for section in parse_sections_list(row_sec)])\n",
    "    concat_standalone_paragraphs = ' '.join(parse_paragraphs_list(row_p))\n",
    "    print(concat_parsed_sections, concat_standalone_paragraphs)\n",
    "    return concat_standalone_paragraphs + \" \" + concat_parsed_sections"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {
    "scrolled": false
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "[(0,\n",
       "  'The mature 26S proteasome consists of at least 33 distinct subunits. Fourteen of them (α1-7 and β1-7) form the 20S core particle (CP), a barrel-shaped structure that encloses three types of peptidase activities (trypsin-like, caspase-like and chymotrypsin-like). The remaining 19 subunits (Rpt1-6, Rpn1-3, 5-13 and 15) constitute the 19S regulatory particle (RP) that caps the CP on one or both ends. Protein substrates destined for proteasomal degradation are captured and processed by the 19S RP before they are threaded into the 20S CP for proteolysis. During this process, the ATPase subunits (Rpt1-6) play key roles in substrate engagement, unfolding, translocation and CP gate opening The 26S proteasome is an essential protein complex responsible for degrading the majority of cellular proteins in eukaryotes'),\n",
       " (1,\n",
       "  'Purely computational approaches have been used to perform virtual screens across multiple mechanisms of action Advances in molecular biology have led to an unprecedented ability to profile the genetic- and pathway-level changes that occur in disease'),\n",
       " (2,\n",
       "  'The zebrafish has recently emerged as a useful vertebrate model system to study neural circuits and behavior, but tools to modulate neurons in freely behaving animals are limited. As a poikilotherm that lives in water, zebrafish are amenable to thermal and pharmacological perturbations. We exploit these properties by using transient receptor potential (TRP) channels to inducibly activate and ablate specific larval zebrafish ( '),\n",
       " (3,\n",
       "  'Somatic copy number alterations (SCNAs), including chromosome arm-level copy changes as well as focal amplifications and deletions, are central events in cancer pathogenesis '),\n",
       " (4,\n",
       "  'Interrupting the flow of information by disconnecting brain regions from one another has been, and remains, a powerful means to investigate interactions between putative components of neural circuits. Previous work from this laboratory has demonstrated that functionally disconnecting rhinal cortex (Rh) and orbitofrontal cortex (OFC) leads to altered estimates of expected value through reward-stimulus association ')]"
      ]
     },
     "execution_count": 7,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "def get_bodys_list(input_dir):\n",
    "    df =   spark.read \\\n",
    "                .format('com.databricks.spark.xml') \\\n",
    "                .options(rowTag='record')\\\n",
    "                .options(rowTag='body')\\\n",
    "                .load(input_dir)\n",
    "    return  df.select(\"sec\", \"p\")\\\n",
    "              .rdd\\\n",
    "              .map(lambda row: build_body(row['sec'], row[\"p\"]))\\\n",
    "              .zipWithIndex()\\\n",
    "              .map(lambda record: (record[1], record[0]))\n",
    "\n",
    "get_bodys_list(input_dir).take(5)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Get article's abstract\n",
    "\n",
    "The abstract of the article is going to be built the same as the body."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {
    "scrolled": false
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "[(0,\n",
       "  'Despite the fundamental importance of proteasomal degradation in cells, little is known about whether and how the 26S proteasome itself is regulated in coordination with various physiological processes. Here we show that the proteasome is dynamically phosphorylated during cell cycle at Thr25 of the 19S subunit Rpt3. CRISPR/Cas9-mediated genome editing, RNA interference and biochemical studies demonstrate that blocking Rpt3-Thr25 phosphorylation markedly impairs proteasome activity and impedes cell proliferation. Through a kinome-wide screen, we have identified dual-specificity tyrosine-regulated kinase 2 (DYRK2) as the primary kinase that phosphorylates Rpt3-Thr25, leading to enhanced substrate translocation and degradation. Importantly, loss of the single phosphorylation of Rpt3-Thr25 or knockout of DYRK2 significantly inhibits tumor formation by proteasome-addicted human breast cancer cells in mice. These findings define an important mechanism for proteasome regulation and demonstrate the biological significance of proteasome phosphorylation in regulating cell proliferation and tumorigenesis. '),\n",
       " (1,\n",
       "  'High-content, image-based screens enable the identification of compounds that induce cellular responses similar to those of known drugs but through different chemical structures or targets. A central challenge in designing phenotypic screens is choosing suitable imaging biomarkers. Here we present a method for systematically identifying optimal reporter cell lines for annotating compound libraries (ORACLs), whose phenotypic profiles most accurately classify a training set of known drugs. We generate a library of fluorescently tagged reporter cell lines, and let analytical criteria determine which among them—the ORACL—best classifies compounds into multiple, diverse drug classes. We demonstrate that an ORACL can functionally annotate large compound libraries across diverse drug classes in a single-pass screen and confirm high prediction accuracy via orthogonal, secondary validation assays. Our approach will increase the efficiency, scale and accuracy of phenotypic screens by maximizing their discriminatory power. '),\n",
       " (2,\n",
       "  'Whole genome analysis approaches are revealing recurrent cancer-associated somatic alterations in non-coding DNA regions. We combined somatic copy number analysis of 12 tumor types with tissue-specific epigenetic profiling to identify significant regions of focal amplification harboring super-enhancers. Copy-number gains of non-coding regions harboring super-enhancers near  '),\n",
       " (3,\n",
       "  'To study how the interaction between orbitofrontal (OFC) and rhinal (Rh) cortices influences the judgment of reward size, we reversibly disconnected these regions using the hM4Di-DREADD (Designer Receptor Exclusively Activated by Designer Drug). Repeated inactivation reduced sensitivity to differences in reward size in two monkeys. Results suggest that retrieval of relative stimulus values from memory appears to depend on interaction between Rh and OFC. '),\n",
       " (4,\n",
       "  ' C o r n e l i a   d e   L a n g e   s y n d r o m e   i s   t h e   p r o t o t y p e   f o r   c o h e s i n o p a t h y   d i s o r d e r s ,   w h i c h   a r e   c h a r a c t e r i z e d   b y   d e f e c t s   i n   c h r o m o s o m e   s e g r e g a t i o n .   K i d n e y   m a l f o r m a t i o n s ,   i n c l u d i n g   n e p h r o g e n i c   r e s t s ,   a r e   c o m m o n   i n   C o r n e l i a   d e   L a n g e   s y n d r o m e .   O n l y   o n e   p o s t - m o r t e m   c a s e   r e p o r t   h a s   d e s c r i b e d   a n   a s s o c i a t i o n   b e t w e e n   W i l m s   t u m o r   a n d   C o r n e l i a   d e   L a n g e   s y n d r o m e .   H e r e ,   w e   d e s c r i b e   t h e   f i r s t   c a s e   o f   a   l i v i n g   c h i l d   w i t h   b o t h   d i s e a s e s . N o n - a n a p l a s t i c   t r i p h a s i c   n e p h r o b l a s t o m a   w a s   d i a g n o s e d   i n   a   p a t i e n t   c a r r y i n g   a   n o t   y e t   r e p o r t e d   m u t a t i o n   i n  ')]"
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "def get_abstract_list(input_dir):\n",
    "    df =   spark.read \\\n",
    "                .format('com.databricks.spark.xml') \\\n",
    "                .options(rowTag='record')\\\n",
    "                .options(rowTag='abstract')\\\n",
    "                .load(input_dir)\n",
    "    return  df.select(\"sec\", \"p\")\\\n",
    "              .rdd\\\n",
    "              .map(lambda row: build_body(row['sec'], row[\"p\"]))\\\n",
    "              .zipWithIndex()\\\n",
    "              .map(lambda record: (record[1], record[0]))\\\n",
    "\n",
    "get_abstract_list(input_dir).take(5)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Get article's categories"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "[(0, ['Article']),\n",
       " (1, ['Article']),\n",
       " (2, ['Article']),\n",
       " (3, ['Article']),\n",
       " (4, ['Article']),\n",
       " (5, ['Case Report']),\n",
       " (6, ['Research Article']),\n",
       " (7, ['Case Report']),\n",
       " (8, ['Research Article']),\n",
       " (9, ['Review']),\n",
       " (10, ['Original Research Article']),\n",
       " (11, ['Research Article']),\n",
       " (12, ['Case Report']),\n",
       " (13, ['Case Report']),\n",
       " (14, ['Articles']),\n",
       " (15, ['Articles']),\n",
       " (16, ['Articles']),\n",
       " (17, ['Articles']),\n",
       " (18, ['Articles']),\n",
       " (19, ['Articles']),\n",
       " (20, ['Articles']),\n",
       " (21, ['Articles']),\n",
       " (22, ['Articles']),\n",
       " (23, ['Articles']),\n",
       " (24, ['Correspondence']),\n",
       " (25, ['Correspondence']),\n",
       " (26, ['Article']),\n",
       " (27, ['Review']),\n",
       " (28, ['Article']),\n",
       " (29, ['Review Article']),\n",
       " (30, ['Research Article']),\n",
       " (31, ['Review Article']),\n",
       " (32, ['Research Article']),\n",
       " (33, ['Research Article']),\n",
       " (34, ['Research Article']),\n",
       " (35, ['Research Article']),\n",
       " (36, ['Research Article']),\n",
       " (37, ['Research Article']),\n",
       " (38, ['Research Article']),\n",
       " (39, ['Research']),\n",
       " (40, ['Research']),\n",
       " (41, ['Research Article']),\n",
       " (42, ['Research']),\n",
       " (43, ['Research Article']),\n",
       " (44, ['Research']),\n",
       " (45, ['Research']),\n",
       " (46, ['Research']),\n",
       " (47, ['Research']),\n",
       " (48, ['Research']),\n",
       " (49, ['Research Article']),\n",
       " (50, ['Research']),\n",
       " (51, ['Research']),\n",
       " (52, ['Original Research']),\n",
       " (53, ['Research']),\n",
       " (54, ['Short Report']),\n",
       " (55, ['Research']),\n",
       " (56, ['Research']),\n",
       " (57, ['Short Report']),\n",
       " (58, ['Study Protocol']),\n",
       " (59, ['Short Report']),\n",
       " (60, ['Research Article']),\n",
       " (61, ['Review']),\n",
       " (62, ['Research']),\n",
       " (63, ['Research Article']),\n",
       " (64, ['Article']),\n",
       " (65, ['Article']),\n",
       " (66, ['Article']),\n",
       " (67, ['Article']),\n",
       " (68, ['Research Article']),\n",
       " (69, ['Research Article']),\n",
       " (70, ['Research Article']),\n",
       " (71, ['Research Article']),\n",
       " (72, ['Article']),\n",
       " (73, ['Article']),\n",
       " (74, ['Article']),\n",
       " (75, ['Article']),\n",
       " (76, ['Article']),\n",
       " (77, ['Article']),\n",
       " (78, ['Article']),\n",
       " (79, ['Article']),\n",
       " (80, ['Article']),\n",
       " (81, ['Article']),\n",
       " (82, ['Research Paper']),\n",
       " (83, ['Article']),\n",
       " (84, ['Short Report']),\n",
       " (85, ['Research Article']),\n",
       " (86, ['Debate']),\n",
       " (87, ['Research']),\n",
       " (88, ['Research Article']),\n",
       " (89, ['Research Articles']),\n",
       " (90, ['Research Articles']),\n",
       " (91, ['Research Articles']),\n",
       " (92, ['Research Articles']),\n",
       " (93, ['Research Articles']),\n",
       " (94, ['Research Articles']),\n",
       " (95, ['Research Articles']),\n",
       " (96, ['Research Articles']),\n",
       " (97, ['Research Articles']),\n",
       " (98, ['Research Articles']),\n",
       " (99, ['Reviews'])]"
      ]
     },
     "execution_count": 9,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "def get_article_categories(row):\n",
    "    categories = []\n",
    "    try: \n",
    "        for subj_group in row['subj-group']:\n",
    "            if type(subj_group['subject']) == str:\n",
    "                categories = [subj_group['subject']]\n",
    "            elif type(subj_group['subject']) == list:\n",
    "                for cat in subj_group['subject']:\n",
    "                    categories.append(cat)\n",
    "    except:\n",
    "        return [row['subj-group']['subject']]\n",
    "    return row['subj-group']\n",
    "    \n",
    "\n",
    "def get_categories_list(input_dir):\n",
    "    df = spark.read \\\n",
    "            .format('com.databricks.spark.xml') \\\n",
    "            .options(rowTag='record')\\\n",
    "            .options(rowTag='metadata')\\\n",
    "            .options(rowTag='article')\\\n",
    "            .options(rowTag='front')\\\n",
    "            .options(rowTag='article-meta')\\\n",
    "            .load(input_dir)\n",
    "    \n",
    "    return  df.select(\"article-categories\")\\\n",
    "              .rdd\\\n",
    "              .map(lambda row: get_article_categories(row['article-categories']))\\\n",
    "              .zipWithIndex()\\\n",
    "              .map(lambda record: (record[1], record[0]))\\\n",
    "\n",
    "get_categories_list(input_dir).take(100)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Get article's title"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "[(0,\n",
       "  'Site-specific Proteasome Phosphorylation Controls Cell Proliferation and Tumorigenesis'),\n",
       " (1,\n",
       "  'Improving drug discovery with high-content phenotypic screens by systematic selection of reporter cell lines'),\n",
       " (2,\n",
       "  'TRP channel mediated neuronal activation and ablation in freely behaving zebrafish'),\n",
       " (3,\n",
       "  'Identification of focally amplified lineage-specific super-enhancers in human epithelial cancers'),\n",
       " (4,\n",
       "  'Disruption of relative reward value by reversible disconnection of orbitofrontal and rhinal cortex using DREADDs in rhesus monkeys')]"
      ]
     },
     "execution_count": 10,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "def get_titles_list(input_dir):\n",
    "    df = spark.read \\\n",
    "            .format('com.databricks.spark.xml') \\\n",
    "            .options(rowTag='record')\\\n",
    "            .options(rowTag='metadata')\\\n",
    "            .options(rowTag='article')\\\n",
    "            .options(rowTag='front')\\\n",
    "            .options(rowTag='article-meta')\\\n",
    "            .options(rowTag='title-group')\\\n",
    "            .load(input_dir)\n",
    "    \n",
    "    return  df.select(\"article-title\")\\\n",
    "              .rdd\\\n",
    "              .map(lambda row: row['article-title'])\\\n",
    "              .zipWithIndex()\\\n",
    "              .map(lambda record: (record[1], record[0]))\\\n",
    "\n",
    "get_titles_list(input_dir).take(5)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Get final dataframe"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {
    "scrolled": false
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>abstract</th>\n",
       "      <th>body</th>\n",
       "      <th>categories</th>\n",
       "      <th>title</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Despite the fundamental importance of proteaso...</td>\n",
       "      <td>The mature 26S proteasome consists of at least...</td>\n",
       "      <td>[Article]</td>\n",
       "      <td>Site-specific Proteasome Phosphorylation Contr...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>High-content, image-based screens enable the i...</td>\n",
       "      <td>Purely computational approaches have been used...</td>\n",
       "      <td>[Article]</td>\n",
       "      <td>Improving drug discovery with high-content phe...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>Whole genome analysis approaches are revealing...</td>\n",
       "      <td>The zebrafish has recently emerged as a useful...</td>\n",
       "      <td>[Article]</td>\n",
       "      <td>TRP channel mediated neuronal activation and a...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>To study how the interaction between orbitofro...</td>\n",
       "      <td>Somatic copy number alterations (SCNAs), inclu...</td>\n",
       "      <td>[Article]</td>\n",
       "      <td>Identification of focally amplified lineage-sp...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>C o r n e l i a   d e   L a n g e   s y n d r...</td>\n",
       "      <td>Interrupting the flow of information by discon...</td>\n",
       "      <td>[Article]</td>\n",
       "      <td>Disruption of relative reward value by reversi...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>109</th>\n",
       "      <td>Pikman et al. demonstrate that the mitochondri...</td>\n",
       "      <td>Cardiolipin (CL) is a dimeric phospholipid tha...</td>\n",
       "      <td>[Article]</td>\n",
       "      <td>Hydrophobic CDR3 residues promote the developm...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>110</th>\n",
       "      <td>Drugs targeting metabolism have formed the bac...</td>\n",
       "      <td>The ability of αβ T cell repertoires to target...</td>\n",
       "      <td>[Article]</td>\n",
       "      <td>The genomic landscape and evolution of endomet...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>111</th>\n",
       "      <td>Antignano, Zaph, and collaborators show that t...</td>\n",
       "      <td>Recent large-scale sequencing studies of prima...</td>\n",
       "      <td>[Article]</td>\n",
       "      <td>Genomic analysis of local variation and recent...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>112</th>\n",
       "      <td>Innate lymphoid cells (ILCs) are emerging as i...</td>\n",
       "      <td>In this study we analysed</td>\n",
       "      <td>[Article]</td>\n",
       "      <td>Recognizing millions of consistently unidentif...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>113</th>\n",
       "      <td>Increased expression of the ER membrane acetyl...</td>\n",
       "      <td>“Untargeted” or “discovery” proteomics approac...</td>\n",
       "      <td>[Article]</td>\n",
       "      <td>Neddylation requires glycyl-tRNA synthetase to...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>114 rows × 4 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                                              abstract  \\\n",
       "0    Despite the fundamental importance of proteaso...   \n",
       "1    High-content, image-based screens enable the i...   \n",
       "2    Whole genome analysis approaches are revealing...   \n",
       "3    To study how the interaction between orbitofro...   \n",
       "4     C o r n e l i a   d e   L a n g e   s y n d r...   \n",
       "..                                                 ...   \n",
       "109  Pikman et al. demonstrate that the mitochondri...   \n",
       "110  Drugs targeting metabolism have formed the bac...   \n",
       "111  Antignano, Zaph, and collaborators show that t...   \n",
       "112  Innate lymphoid cells (ILCs) are emerging as i...   \n",
       "113  Increased expression of the ER membrane acetyl...   \n",
       "\n",
       "                                                  body categories  \\\n",
       "0    The mature 26S proteasome consists of at least...  [Article]   \n",
       "1    Purely computational approaches have been used...  [Article]   \n",
       "2    The zebrafish has recently emerged as a useful...  [Article]   \n",
       "3    Somatic copy number alterations (SCNAs), inclu...  [Article]   \n",
       "4    Interrupting the flow of information by discon...  [Article]   \n",
       "..                                                 ...        ...   \n",
       "109  Cardiolipin (CL) is a dimeric phospholipid tha...  [Article]   \n",
       "110  The ability of αβ T cell repertoires to target...  [Article]   \n",
       "111  Recent large-scale sequencing studies of prima...  [Article]   \n",
       "112                        In this study we analysed    [Article]   \n",
       "113  “Untargeted” or “discovery” proteomics approac...  [Article]   \n",
       "\n",
       "                                                 title  \n",
       "0    Site-specific Proteasome Phosphorylation Contr...  \n",
       "1    Improving drug discovery with high-content phe...  \n",
       "2    TRP channel mediated neuronal activation and a...  \n",
       "3    Identification of focally amplified lineage-sp...  \n",
       "4    Disruption of relative reward value by reversi...  \n",
       "..                                                 ...  \n",
       "109  Hydrophobic CDR3 residues promote the developm...  \n",
       "110  The genomic landscape and evolution of endomet...  \n",
       "111  Genomic analysis of local variation and recent...  \n",
       "112  Recognizing millions of consistently unidentif...  \n",
       "113  Neddylation requires glycyl-tRNA synthetase to...  \n",
       "\n",
       "[114 rows x 4 columns]"
      ]
     },
     "execution_count": 11,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "def get_final_rdd():\n",
    "    final_rdd = get_abstract_list(input_dir)\n",
    "    final_rdd = final_rdd.join(get_bodys_list(input_dir))\n",
    "    final_rdd = final_rdd.join(get_categories_list(input_dir))\n",
    "    final_rdd = final_rdd.join(get_titles_list(input_dir))\n",
    "    return final_rdd.sortByKey()\n",
    "\n",
    "def get_final_to_pandas():\n",
    "    return get_final_rdd().map(lambda record: (record[1][0][0][0], record[1][0][0][1], record[1][0][1], record[1][1]))\\\n",
    "                          .toDF([\"abstract\", \"body\", \"categories\", \"title\"])\n",
    "\n",
    "get_final_to_pandas().toPandas()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Get number of pages"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "[(0, 'Not specified'),\n",
       " (1, 'Not specified'),\n",
       " (2, 'Not specified'),\n",
       " (3, 'Not specified'),\n",
       " (4, 'Not specified'),\n",
       " (5, 'Not specified'),\n",
       " (6, 'Not specified'),\n",
       " (7, 'Not specified'),\n",
       " (8, 'Not specified'),\n",
       " (9, 'Not specified'),\n",
       " (10, 'Not specified'),\n",
       " (11, 'Not specified'),\n",
       " (12, 'Not specified'),\n",
       " (13, 'Not specified'),\n",
       " (14, 'Not specified'),\n",
       " (15, 'Not specified'),\n",
       " (16, 'Not specified'),\n",
       " (17, 'Not specified'),\n",
       " (18, 'Not specified'),\n",
       " (19, 'Not specified'),\n",
       " (20, 'Not specified'),\n",
       " (21, 'Not specified'),\n",
       " (22, 'Not specified'),\n",
       " (23, 'Not specified'),\n",
       " (24, 'Not specified'),\n",
       " (25, 'Not specified'),\n",
       " (26, 'Not specified'),\n",
       " (27, 'Not specified'),\n",
       " (28, 'Not specified'),\n",
       " (29, 'Not specified'),\n",
       " (30, 'Not specified'),\n",
       " (31, 'Not specified'),\n",
       " (32, 'Not specified'),\n",
       " (33, 'Not specified'),\n",
       " (34, 'Not specified'),\n",
       " (35, 7),\n",
       " (36, 'Not specified'),\n",
       " (37, 'Not specified'),\n",
       " (38, 'Not specified'),\n",
       " (39, 'Not specified'),\n",
       " (40, 'Not specified'),\n",
       " (41, 'Not specified'),\n",
       " (42, 'Not specified'),\n",
       " (43, 'Not specified'),\n",
       " (44, 'Not specified'),\n",
       " (45, 'Not specified'),\n",
       " (46, 'Not specified'),\n",
       " (47, 'Not specified'),\n",
       " (48, 'Not specified'),\n",
       " (49, 'Not specified'),\n",
       " (50, 'Not specified'),\n",
       " (51, 'Not specified'),\n",
       " (52, 'Not specified'),\n",
       " (53, 'Not specified'),\n",
       " (54, 'Not specified'),\n",
       " (55, 'Not specified'),\n",
       " (56, 'Not specified'),\n",
       " (57, 'Not specified'),\n",
       " (58, 'Not specified'),\n",
       " (59, 'Not specified'),\n",
       " (60, 'Not specified'),\n",
       " (61, 'Not specified'),\n",
       " (62, 'Not specified'),\n",
       " (63, 'Not specified'),\n",
       " (64, 'Not specified'),\n",
       " (65, 'Not specified'),\n",
       " (66, 'Not specified'),\n",
       " (67, 'Not specified'),\n",
       " (68, 'Not specified'),\n",
       " (69, 'Not specified'),\n",
       " (70, 'Not specified'),\n",
       " (71, 'Not specified'),\n",
       " (72, 'Not specified'),\n",
       " (73, 'Not specified'),\n",
       " (74, 'Not specified'),\n",
       " (75, 'Not specified'),\n",
       " (76, 'Not specified'),\n",
       " (77, 'Not specified'),\n",
       " (78, 'Not specified'),\n",
       " (79, 'Not specified'),\n",
       " (80, 'Not specified'),\n",
       " (81, 'Not specified'),\n",
       " (82, 'Not specified'),\n",
       " (83, 'Not specified'),\n",
       " (84, 'Not specified'),\n",
       " (85, 'Not specified'),\n",
       " (86, 'Not specified'),\n",
       " (87, 'Not specified'),\n",
       " (88, 'Not specified'),\n",
       " (89, 'Not specified'),\n",
       " (90, 'Not specified'),\n",
       " (91, 'Not specified'),\n",
       " (92, 'Not specified'),\n",
       " (93, 'Not specified'),\n",
       " (94, 'Not specified'),\n",
       " (95, 'Not specified'),\n",
       " (96, 'Not specified'),\n",
       " (97, 'Not specified'),\n",
       " (98, 'Not specified'),\n",
       " (99, 'Not specified'),\n",
       " (100, 'Not specified'),\n",
       " (101, 'Not specified'),\n",
       " (102, 'Not specified'),\n",
       " (103, 'Not specified'),\n",
       " (104, 'Not specified'),\n",
       " (105, 'Not specified'),\n",
       " (106, 'Not specified'),\n",
       " (107, 'Not specified'),\n",
       " (108, 'Not specified'),\n",
       " (109, 'Not specified'),\n",
       " (110, 'Not specified'),\n",
       " (111, 'Not specified'),\n",
       " (112, 'Not specified'),\n",
       " (113, 'Not specified')]"
      ]
     },
     "execution_count": 12,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "def get_page_counts(row):\n",
    "    try:\n",
    "        return row['page-count']._count\n",
    "    except:\n",
    "        return 'Not specified'\n",
    "    \n",
    "def get_number_of_pages_list(input_dir):\n",
    "    df = spark.read \\\n",
    "            .format('com.databricks.spark.xml') \\\n",
    "            .options(rowTag='record')\\\n",
    "            .options(rowTag='metadata')\\\n",
    "            .options(rowTag='article')\\\n",
    "            .load(input_dir)\n",
    "    \n",
    "    return  df.select(\"counts\")\\\n",
    "              .rdd\\\n",
    "              .map(lambda row: get_page_counts(row['counts']))\\\n",
    "              .zipWithIndex()\\\n",
    "              .map(lambda record: (record[1], record[0]))\\\n",
    "\n",
    "get_number_of_pages_list(input_dir).collect()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Get authors list"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "[(0,\n",
       "  ['Xing Guo',\n",
       "   'Xiaorong Wang',\n",
       "   'Zhiping Wang',\n",
       "   'Sourav Banerjee',\n",
       "   'Jing Yang',\n",
       "   'Lan Huang',\n",
       "   'Jack E. Dixon']),\n",
       " (1,\n",
       "  ['Jungseog Kang',\n",
       "   'Chien-Hsiang Hsu',\n",
       "   'Qi Wu',\n",
       "   'Shanshan Liu',\n",
       "   'Adam D. Coster',\n",
       "   'Bruce A. Posner',\n",
       "   'Steven J. Altschuler',\n",
       "   'Lani F. Wu']),\n",
       " (2,\n",
       "  ['Shijia Chen',\n",
       "   'Cindy N. Chiu',\n",
       "   'Kimberly L. McArthur',\n",
       "   'Joseph R. Fetcho',\n",
       "   'David A. Prober']),\n",
       " (3,\n",
       "  ['Xiaoyang Zhang',\n",
       "   'Peter S. Choi',\n",
       "   'Joshua M. Francis',\n",
       "   'Marcin Imielinski',\n",
       "   'Hideo Watanabe',\n",
       "   'Andrew D. Cherniack',\n",
       "   'Matthew Meyerson']),\n",
       " (4,\n",
       "  ['Mark A G Eldridge',\n",
       "   'Walter Lerchner',\n",
       "   'Richard C Saunders',\n",
       "   'Hiroyuki Kaneko',\n",
       "   'Kristopher W Krausz',\n",
       "   'Frank J Gonzalez',\n",
       "   'Bin Ji',\n",
       "   'Makoto Higuchi',\n",
       "   'Takafumi Minamimoto',\n",
       "   'Barry J Richmond'])]"
      ]
     },
     "execution_count": 13,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "def get_authors(row):\n",
    "    authors = []\n",
    "    try:\n",
    "        for contrib in row['contrib']:\n",
    "            print(contrib)\n",
    "            if contrib['_contrib-type'] == 'author':\n",
    "                authors.append(contrib['name']['given-names'] + ' ' + contrib['name']['surname'])\n",
    "        return authors\n",
    "    except:\n",
    "        return 'Not specified'\n",
    "    \n",
    "def get_authors_list(input_dir):\n",
    "    df = spark.read \\\n",
    "            .format('com.databricks.spark.xml') \\\n",
    "            .options(rowTag='record')\\\n",
    "            .options(rowTag='metadata')\\\n",
    "            .options(rowTag='article')\\\n",
    "            .options(rowTag='front')\\\n",
    "            .options(rowTag='article-meta')\\\n",
    "            .load(input_dir)\n",
    "    \n",
    "    return  df.select(\"contrib-group\")\\\n",
    "              .rdd\\\n",
    "              .map(lambda row: get_authors(row['contrib-group']))\\\n",
    "              .zipWithIndex()\\\n",
    "              .map(lambda record: (record[1], record[0]))\\\n",
    "\n",
    "get_authors_list(input_dir).take(5)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Get author's affiliation"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {
    "scrolled": false
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "[(0, 'Not specified'),\n",
       " (1, 'Not specified'),\n",
       " (2, 'Not specified'),\n",
       " (3, 'Not specified'),\n",
       " (4, 'Not specified'),\n",
       " (5, 'Not specified'),\n",
       " (6, 'Not specified'),\n",
       " (7, 'Not specified'),\n",
       " (8, 'Not specified'),\n",
       " (9, 'Not specified'),\n",
       " (10, 'Not specified'),\n",
       " (11, 'Not specified'),\n",
       " (12, 'Not specified'),\n",
       " (13, 'Not specified'),\n",
       " (14, 'Not specified'),\n",
       " (15, 'Not specified'),\n",
       " (16, 'Not specified'),\n",
       " (17, 'Not specified'),\n",
       " (18, 'Not specified'),\n",
       " (19, 'Not specified'),\n",
       " (20, 'Not specified'),\n",
       " (21, 'Not specified'),\n",
       " (22, 'Not specified'),\n",
       " (23, 'Not specified'),\n",
       " (24, 'Not specified'),\n",
       " (25, 'Not specified'),\n",
       " (26, 'Not specified'),\n",
       " (27, 'Not specified'),\n",
       " (28, 'Not specified'),\n",
       " (29, 'Not specified'),\n",
       " (30, 'Not specified'),\n",
       " (31, 'Not specified'),\n",
       " (32, 'Not specified'),\n",
       " (33, 'Not specified'),\n",
       " (34, 'Not specified'),\n",
       " (35, 'Not specified'),\n",
       " (36, 'Not specified'),\n",
       " (37, 'Not specified'),\n",
       " (38, 'Not specified'),\n",
       " (39, 'Not specified'),\n",
       " (40, 'Not specified'),\n",
       " (41, 'Not specified'),\n",
       " (42, 'Not specified'),\n",
       " (43, 'Not specified'),\n",
       " (44, 'Not specified'),\n",
       " (45, 'Not specified'),\n",
       " (46, 'Not specified'),\n",
       " (47, 'Not specified'),\n",
       " (48, 'Not specified'),\n",
       " (49, 'Not specified'),\n",
       " (50, 'Not specified'),\n",
       " (51, 'Not specified'),\n",
       " (52, 'Not specified'),\n",
       " (53, 'Not specified'),\n",
       " (54, 'Not specified'),\n",
       " (55, 'Not specified'),\n",
       " (56, 'Not specified'),\n",
       " (57, 'Not specified'),\n",
       " (58, 'Not specified'),\n",
       " (59, 'Not specified'),\n",
       " (60, 'Not specified'),\n",
       " (61, 'Not specified'),\n",
       " (62, 'Not specified'),\n",
       " (63, 'Not specified'),\n",
       " (64, 'Not specified'),\n",
       " (65, 'Not specified'),\n",
       " (66, 'Not specified'),\n",
       " (67, 'Not specified'),\n",
       " (68, 'Not specified'),\n",
       " (69, 'Not specified'),\n",
       " (70, 'Not specified'),\n",
       " (71, 'Not specified'),\n",
       " (72, 'Not specified'),\n",
       " (73, 'Not specified'),\n",
       " (74, 'Not specified'),\n",
       " (75, 'Not specified'),\n",
       " (76, 'Not specified'),\n",
       " (77, 'Not specified'),\n",
       " (78, 'Not specified'),\n",
       " (79, 'Not specified'),\n",
       " (80, 'Not specified'),\n",
       " (81, 'Not specified'),\n",
       " (82, 'Not specified'),\n",
       " (83, 'Not specified'),\n",
       " (84, 'Not specified'),\n",
       " (85, 'Not specified'),\n",
       " (86, 'Not specified'),\n",
       " (87, 'Not specified'),\n",
       " (88, 'Not specified'),\n",
       " (89, 'Not specified'),\n",
       " (90, 'Not specified'),\n",
       " (91, 'Not specified'),\n",
       " (92, 'Not specified'),\n",
       " (93, 'Not specified'),\n",
       " (94, 'Not specified'),\n",
       " (95, 'Not specified'),\n",
       " (96, 'Not specified'),\n",
       " (97, 'Not specified'),\n",
       " (98, 'Not specified'),\n",
       " (99, 'Not specified'),\n",
       " (100, 'Not specified'),\n",
       " (101, 'Not specified'),\n",
       " (102, 'Not specified'),\n",
       " (103, 'Not specified'),\n",
       " (104, 'Not specified'),\n",
       " (105, 'Not specified'),\n",
       " (106, 'Not specified'),\n",
       " (107, 'Not specified'),\n",
       " (108, 'Not specified'),\n",
       " (109, 'Not specified'),\n",
       " (110, 'Not specified'),\n",
       " (111, 'Not specified'),\n",
       " (112, 'Not specified'),\n",
       " (113, 'Not specified')]"
      ]
     },
     "execution_count": 14,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "def get_affiliations(row):\n",
    "    affiliations = dict()\n",
    "    affiliations_for_authors = []\n",
    "    try:\n",
    "        for aff in row['aff']:\n",
    "            affiliations[aff['_id']] = (aff['country'], aff['institution'])\n",
    "            \n",
    "        for contrib in row['contrib']:\n",
    "            if contrib['_contrib-type'] == 'author':\n",
    "                ids_aff = [aff['_rid'] for aff in contrib['xref']]\n",
    "                affiliations_for_authors.append([])\n",
    "                \n",
    "                for id_aff in ids_aff:\n",
    "                    affiliations_for_authors[-1].append(affiliations[id_aff])\n",
    "        return affiliations_for_authors\n",
    "    except:\n",
    "        return 'Not specified'\n",
    "    \n",
    "def get_affiliations_list(input_dir):\n",
    "    df = spark.read \\\n",
    "            .format('com.databricks.spark.xml') \\\n",
    "            .options(rowTag='record')\\\n",
    "            .options(rowTag='metadata')\\\n",
    "            .options(rowTag='article')\\\n",
    "            .options(rowTag='front')\\\n",
    "            .options(rowTag='article-meta')\\\n",
    "            .load(input_dir)\n",
    "    \n",
    "    return  df.select(\"contrib-group\")\\\n",
    "              .rdd\\\n",
    "              .map(lambda row: get_affiliations(row['contrib-group']))\\\n",
    "              .zipWithIndex()\\\n",
    "              .map(lambda record: (record[1], record[0]))\\\n",
    "\n",
    "get_affiliations_list(input_dir).collect()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Get number of figures"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "[(0, 'Not specified'),\n",
       " (1, 'Not specified'),\n",
       " (2, 'Not specified'),\n",
       " (3, 'Not specified'),\n",
       " (4, 'Not specified'),\n",
       " (5, 'Not specified'),\n",
       " (6, 'Not specified'),\n",
       " (7, 'Not specified'),\n",
       " (8, 'Not specified'),\n",
       " (9, 'Not specified'),\n",
       " (10, 'Not specified'),\n",
       " (11, 'Not specified'),\n",
       " (12, 'Not specified'),\n",
       " (13, 'Not specified'),\n",
       " (14, 'Not specified'),\n",
       " (15, 1),\n",
       " (16, 'Not specified'),\n",
       " (17, 1),\n",
       " (18, 'Not specified'),\n",
       " (19, 'Not specified'),\n",
       " (20, 'Not specified'),\n",
       " (21, 1),\n",
       " (22, 'Not specified'),\n",
       " (23, 'Not specified'),\n",
       " (24, 'Not specified'),\n",
       " (25, 'Not specified'),\n",
       " (26, 'Not specified'),\n",
       " (27, 'Not specified'),\n",
       " (28, 'Not specified'),\n",
       " (29, 'Not specified'),\n",
       " (30, 'Not specified'),\n",
       " (31, 'Not specified'),\n",
       " (32, 1),\n",
       " (33, 'Not specified'),\n",
       " (34, 'Not specified'),\n",
       " (35, 'Not specified'),\n",
       " (36, 'Not specified'),\n",
       " (37, 'Not specified'),\n",
       " (38, 'Not specified'),\n",
       " (39, 'Not specified'),\n",
       " (40, 'Not specified'),\n",
       " (41, 'Not specified'),\n",
       " (42, 'Not specified'),\n",
       " (43, 'Not specified'),\n",
       " (44, 'Not specified'),\n",
       " (45, 'Not specified'),\n",
       " (46, 'Not specified'),\n",
       " (47, 'Not specified'),\n",
       " (48, 'Not specified'),\n",
       " (49, 'Not specified'),\n",
       " (50, 'Not specified'),\n",
       " (51, 'Not specified'),\n",
       " (52, 'Not specified'),\n",
       " (53, 'Not specified'),\n",
       " (54, 'Not specified'),\n",
       " (55, 'Not specified'),\n",
       " (56, 'Not specified'),\n",
       " (57, 'Not specified'),\n",
       " (58, 'Not specified'),\n",
       " (59, 'Not specified'),\n",
       " (60, 'Not specified'),\n",
       " (61, 'Not specified'),\n",
       " (62, 'Not specified'),\n",
       " (63, 'Not specified'),\n",
       " (64, 'Not specified'),\n",
       " (65, 'Not specified'),\n",
       " (66, 'Not specified'),\n",
       " (67, 'Not specified'),\n",
       " (68, 'Not specified'),\n",
       " (69, 'Not specified'),\n",
       " (70, 'Not specified'),\n",
       " (71, 'Not specified'),\n",
       " (72, 'Not specified'),\n",
       " (73, 'Not specified'),\n",
       " (74, 'Not specified'),\n",
       " (75, 'Not specified'),\n",
       " (76, 'Not specified'),\n",
       " (77, 'Not specified'),\n",
       " (78, 'Not specified'),\n",
       " (79, 'Not specified'),\n",
       " (80, 'Not specified'),\n",
       " (81, 'Not specified'),\n",
       " (82, 'Not specified'),\n",
       " (83, 'Not specified'),\n",
       " (84, 'Not specified'),\n",
       " (85, 'Not specified'),\n",
       " (86, 'Not specified'),\n",
       " (87, 'Not specified'),\n",
       " (88, 'Not specified'),\n",
       " (89, 'Not specified'),\n",
       " (90, 'Not specified'),\n",
       " (91, 'Not specified'),\n",
       " (92, 'Not specified'),\n",
       " (93, 'Not specified'),\n",
       " (94, 'Not specified'),\n",
       " (95, 'Not specified'),\n",
       " (96, 'Not specified'),\n",
       " (97, 'Not specified'),\n",
       " (98, 'Not specified'),\n",
       " (99, 'Not specified'),\n",
       " (100, 'Not specified'),\n",
       " (101, 'Not specified'),\n",
       " (102, 'Not specified'),\n",
       " (103, 'Not specified'),\n",
       " (104, 'Not specified'),\n",
       " (105, 'Not specified'),\n",
       " (106, 'Not specified'),\n",
       " (107, 'Not specified'),\n",
       " (108, 'Not specified'),\n",
       " (109, 'Not specified'),\n",
       " (110, 'Not specified'),\n",
       " (111, 'Not specified'),\n",
       " (112, 'Not specified'),\n",
       " (113, 'Not specified')]"
      ]
     },
     "execution_count": 15,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "def get_fig_no(row):\n",
    "    try:\n",
    "        if row['fig'] != None:\n",
    "            if type(row['fig']) == list:\n",
    "                return len(row['fig'])\n",
    "            return 1\n",
    "        else:\n",
    "            return 'Not specified' \n",
    "    except:\n",
    "        return 'Not specified'\n",
    "\n",
    "def get_number_of_figures(input_dir):\n",
    "    df = spark.read \\\n",
    "            .format('com.databricks.spark.xml') \\\n",
    "            .options(rowTag='record')\\\n",
    "            .load(input_dir)\n",
    "    \n",
    "    return  df.select('fig')\\\n",
    "              .rdd\\\n",
    "              .map(lambda row: get_fig_no(row))\\\n",
    "              .zipWithIndex()\\\n",
    "              .map(lambda record: (record[1], record[0]))\\\n",
    "\n",
    "get_number_of_figures(input_dir).collect()"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.3"
  },
  "toc": {
   "base_numbering": 1,
   "nav_menu": {},
   "number_sections": false,
   "sideBar": true,
   "skip_h1_title": false,
   "title_cell": "Data parsing",
   "title_sidebar": "Contents",
   "toc_cell": true,
   "toc_position": {
    "height": "calc(100% - 180px)",
    "left": "10px",
    "top": "150px",
    "width": "373.333px"
   },
   "toc_section_display": true,
   "toc_window_display": false
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
