{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {
    "toc": true
   },
   "source": [
    "<h1>Data parsing<span class=\"tocSkip\"></span></h1>\n",
    "<div class=\"toc\"><ul class=\"toc-item\"><li><span><a href=\"#Import-dependencies\" data-toc-modified-id=\"Import-dependencies-1\">Import dependencies</a></span></li><li><span><a href=\"#Load-data\" data-toc-modified-id=\"Load-data-2\">Load data</a></span></li></ul></div>"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Import dependencies"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "import os, shutil\n",
    "import warnings\n",
    "import random\n",
    "warnings.simplefilter(\"ignore\")\n",
    "\n",
    "from shutil import copyfile\n",
    "from distutils.dir_util import copy_tree"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Load data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "file_path = \"../datasets/2016_testing_df/\"\n",
    "\n",
    "df = sqlContext.read.format('com.databricks.spark.xml').option(\"rowTag\", \"record\")\\\n",
    "    .load(file_path)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "103"
      ]
     },
     "execution_count": 2,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df.rdd.count()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>abstract</th>\n",
       "      <th>body</th>\n",
       "      <th>custom-meta-group</th>\n",
       "      <th>elocation-id</th>\n",
       "      <th>ext-link</th>\n",
       "      <th>fig</th>\n",
       "      <th>funding-group</th>\n",
       "      <th>graphic</th>\n",
       "      <th>header</th>\n",
       "      <th>history</th>\n",
       "      <th>...</th>\n",
       "      <th>permissions</th>\n",
       "      <th>pub-date</th>\n",
       "      <th>sec</th>\n",
       "      <th>self-uri</th>\n",
       "      <th>sub</th>\n",
       "      <th>sup</th>\n",
       "      <th>td</th>\n",
       "      <th>tr</th>\n",
       "      <th>volume</th>\n",
       "      <th>xref</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>NaN</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>(2016-06-01, oai:pubmedcentral.nih.gov:4868590...</td>\n",
       "      <td>None</td>\n",
       "      <td>...</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>NaN</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>None</td>\n",
       "      <td>([et al], None, (None, None, [et al], None, (P...</td>\n",
       "      <td>None</td>\n",
       "      <td>NaN</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>(2016-06-01, oai:pubmedcentral.nih.gov:4869041...</td>\n",
       "      <td>None</td>\n",
       "      <td>...</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>NaN</td>\n",
       "      <td>[(Verdejo-Garcia , bibr, bib46)]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>NaN</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>(2016-06-01, oai:pubmedcentral.nih.gov:4869059...</td>\n",
       "      <td>None</td>\n",
       "      <td>...</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>NaN</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>NaN</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>(2016-06-01, oai:pubmedcentral.nih.gov:4869150...</td>\n",
       "      <td>None</td>\n",
       "      <td>...</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>[(s0020, None, None, [htfr], 2.2, [EGFP-CEM-NK...</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>NaN</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>NaN</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>(2016-06-01, oai:pubmedcentral.nih.gov:4869590...</td>\n",
       "      <td>None</td>\n",
       "      <td>...</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>NaN</td>\n",
       "      <td>[(Debette et al., 2011, bibr, bib10), (van Dij...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows × 26 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "  abstract                                               body  \\\n",
       "0     None                                               None   \n",
       "1     None  ([et al], None, (None, None, [et al], None, (P...   \n",
       "2     None                                               None   \n",
       "3     None                                               None   \n",
       "4     None                                               None   \n",
       "\n",
       "  custom-meta-group  elocation-id ext-link   fig funding-group graphic  \\\n",
       "0              None           NaN     None  None          None    None   \n",
       "1              None           NaN     None  None          None    None   \n",
       "2              None           NaN     None  None          None    None   \n",
       "3              None           NaN     None  None          None    None   \n",
       "4              None           NaN     None  None          None    None   \n",
       "\n",
       "                                              header history  ...  \\\n",
       "0  (2016-06-01, oai:pubmedcentral.nih.gov:4868590...    None  ...   \n",
       "1  (2016-06-01, oai:pubmedcentral.nih.gov:4869041...    None  ...   \n",
       "2  (2016-06-01, oai:pubmedcentral.nih.gov:4869059...    None  ...   \n",
       "3  (2016-06-01, oai:pubmedcentral.nih.gov:4869150...    None  ...   \n",
       "4  (2016-06-01, oai:pubmedcentral.nih.gov:4869590...    None  ...   \n",
       "\n",
       "   permissions pub-date                                                sec  \\\n",
       "0         None     None                                               None   \n",
       "1         None     None                                               None   \n",
       "2         None     None                                               None   \n",
       "3         None     None  [(s0020, None, None, [htfr], 2.2, [EGFP-CEM-NK...   \n",
       "4         None     None                                               None   \n",
       "\n",
       "  self-uri   sub   sup    td    tr volume  \\\n",
       "0     None  None  None  None  None    NaN   \n",
       "1     None  None  None  None  None    NaN   \n",
       "2     None  None  None  None  None    NaN   \n",
       "3     None  None  None  None  None    NaN   \n",
       "4     None  None  None  None  None    NaN   \n",
       "\n",
       "                                                xref  \n",
       "0                                               None  \n",
       "1                   [(Verdejo-Garcia , bibr, bib46)]  \n",
       "2                                               None  \n",
       "3                                               None  \n",
       "4  [(Debette et al., 2011, bibr, bib10), (van Dij...  \n",
       "\n",
       "[5 rows x 26 columns]"
      ]
     },
     "execution_count": 3,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df.toPandas().head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "['abstract',\n",
       " 'body',\n",
       " 'custom-meta-group',\n",
       " 'elocation-id',\n",
       " 'ext-link',\n",
       " 'fig',\n",
       " 'funding-group',\n",
       " 'graphic',\n",
       " 'header',\n",
       " 'history',\n",
       " 'issue',\n",
       " 'issue-title',\n",
       " 'italic',\n",
       " 'kwd-group',\n",
       " 'metadata',\n",
       " 'p',\n",
       " 'permissions',\n",
       " 'pub-date',\n",
       " 'sec',\n",
       " 'self-uri',\n",
       " 'sub',\n",
       " 'sup',\n",
       " 'td',\n",
       " 'tr',\n",
       " 'volume',\n",
       " 'xref']"
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "list(df.toPandas())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "sqlContext.read.format('com.databricks.spark.xml').option(\"rowTag\", \"record\")\\\n",
    "    .load(file_path)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>metadata</th>\n",
       "      <th>fig</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>&lt;article&gt;\\n  &lt;front&gt;\\n    &lt;journal-meta&gt;\\n    ...</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>&lt;article&gt;\\n  &lt;front&gt;\\n    &lt;journal-meta&gt;\\n    ...</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>&lt;article&gt;\\n  &lt;front&gt;\\n    &lt;journal-meta&gt;\\n    ...</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>&lt;article&gt;\\n  &lt;front&gt;\\n    &lt;journal-meta&gt;\\n    ...</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>&lt;article&gt;\\n  &lt;front&gt;\\n    &lt;journal-meta&gt;\\n    ...</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>98</th>\n",
       "      <td>&lt;article&gt;\\n  &lt;front&gt;\\n    &lt;journal-meta&gt;\\n    ...</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>99</th>\n",
       "      <td>&lt;article&gt;\\n  &lt;front&gt;\\n    &lt;journal-meta&gt;\\n    ...</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>100</th>\n",
       "      <td>&lt;article&gt;\\n  &lt;front&gt;\\n    &lt;journal-meta&gt;\\n    ...</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>101</th>\n",
       "      <td>&lt;article&gt;\\n  &lt;front&gt;\\n    &lt;journal-meta&gt;\\n    ...</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>102</th>\n",
       "      <td>&lt;article&gt;\\n  &lt;front&gt;\\n    &lt;journal-meta&gt;\\n    ...</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>103 rows × 2 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                                              metadata   fig\n",
       "0    <article>\\n  <front>\\n    <journal-meta>\\n    ...  None\n",
       "1    <article>\\n  <front>\\n    <journal-meta>\\n    ...  None\n",
       "2    <article>\\n  <front>\\n    <journal-meta>\\n    ...  None\n",
       "3    <article>\\n  <front>\\n    <journal-meta>\\n    ...  None\n",
       "4    <article>\\n  <front>\\n    <journal-meta>\\n    ...  None\n",
       "..                                                 ...   ...\n",
       "98   <article>\\n  <front>\\n    <journal-meta>\\n    ...  None\n",
       "99   <article>\\n  <front>\\n    <journal-meta>\\n    ...  None\n",
       "100  <article>\\n  <front>\\n    <journal-meta>\\n    ...  None\n",
       "101  <article>\\n  <front>\\n    <journal-meta>\\n    ...  None\n",
       "102  <article>\\n  <front>\\n    <journal-meta>\\n    ...  None\n",
       "\n",
       "[103 rows x 2 columns]"
      ]
     },
     "execution_count": 14,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "from pyspark.sql.types import *\n",
    "\n",
    "customSchema = StructType([\n",
    "    StructField(\"metadata\", StringType(), True),\n",
    "    StructField(\"fig\", StringType(), True)])\n",
    "\n",
    "df = spark.read \\\n",
    "    .format('xml') \\\n",
    "    .options(rowTag='record') \\\n",
    "    .load(file_path, schema = customSchema)\n",
    "\n",
    "df.select(\"metadata\", \"fig\").toPandas()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 151,
   "metadata": {
    "scrolled": false
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "['abstract',\n",
       " 'body',\n",
       " 'custom-meta-group',\n",
       " 'elocation-id',\n",
       " 'ext-link',\n",
       " 'fig',\n",
       " 'funding-group',\n",
       " 'graphic',\n",
       " 'header',\n",
       " 'history',\n",
       " 'issue',\n",
       " 'issue-title',\n",
       " 'italic',\n",
       " 'kwd-group',\n",
       " 'metadata',\n",
       " 'p',\n",
       " 'permissions',\n",
       " 'pub-date',\n",
       " 'sec',\n",
       " 'self-uri',\n",
       " 'sub',\n",
       " 'sup',\n",
       " 'td',\n",
       " 'tr',\n",
       " 'volume',\n",
       " 'xref']"
      ]
     },
     "execution_count": 151,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df = spark.read \\\n",
    "    .format('xml') \\\n",
    "    .options(rowTag='record') \\\n",
    "    .load(file_path)\n",
    "list(df.toPandas())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 78,
   "metadata": {
    "scrolled": false
   },
   "outputs": [
    {
     "ename": "TypeError",
     "evalue": "'JavaPackage' object is not callable",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mTypeError\u001b[0m                                 Traceback (most recent call last)",
      "\u001b[0;32m<ipython-input-78-67586bc10a76>\u001b[0m in \u001b[0;36m<module>\u001b[0;34m\u001b[0m\n\u001b[1;32m     31\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m     32\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m---> 33\u001b[0;31m \u001b[0mnew_df\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mdf\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mselect\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mfrom_xml\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mdf\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0;34m'body'\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mschema\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m{\u001b[0m\u001b[0;34m\"rowTag\"\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\"article\"\u001b[0m\u001b[0;34m}\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m",
      "\u001b[0;32m<ipython-input-78-67586bc10a76>\u001b[0m in \u001b[0;36mfrom_xml\u001b[0;34m(col, schema, options)\u001b[0m\n\u001b[1;32m     27\u001b[0m     \u001b[0mscala_datatype\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mspark\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m_jsparkSession\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mparseDataType\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mschema\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mjson\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m     28\u001b[0m     \u001b[0mscala_options\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0msc\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m_jvm\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mPythonUtils\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mtoScalaMap\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0moptions\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m---> 29\u001b[0;31m     \u001b[0mjc\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0msc\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m_jvm\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mcom\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mdatabricks\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mspark\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mxml\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mfunctions\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mfrom_xml\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mcol\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m_jc\u001b[0m \u001b[0;32mif\u001b[0m \u001b[0misinstance\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mcol\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mColumn\u001b[0m\u001b[0;34m)\u001b[0m \u001b[0;32melse\u001b[0m \u001b[0mcol\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mscala_datatype\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mscala_options\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m     30\u001b[0m     \u001b[0;32mreturn\u001b[0m \u001b[0mColumn\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mjc\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m     31\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;31mTypeError\u001b[0m: 'JavaPackage' object is not callable"
     ]
    }
   ],
   "source": [
    "from pyspark.sql.column import Column\n",
    "\n",
    "\n",
    "df = spark.read \\\n",
    "    .format('com.databricks.spark.xml') \\\n",
    "    .options(rowTag='record')\\\n",
    "    .options(rowTag='metadata')\\\n",
    "    .options(rowTag='article')\\\n",
    "    .options(rowTag='body')\\\n",
    "    .load(file_path)\n",
    "# schema = df.schema\n",
    "\n",
    "# df = spark.read \\\n",
    "#     .format('com.databricks.spark.xml') \\\n",
    "#     .options(rowTag='record')\\\n",
    "#     .options(rowTag='metadata')\\\n",
    "#     .options(rowTag='article')\\\n",
    "#     .load(file_path)\n",
    "\n",
    "# # df.toPandas()\n",
    "\n",
    "# from pyspark.sql.column import Column, _to_java_column\n",
    "# from pyspark.sql.types import _parse_datatype_json_string\n",
    "# from pyspark.sql.types import *\n",
    "\n",
    "# def from_xml(col, schema, options={}):\n",
    "#     scala_datatype = spark._jsparkSession.parseDataType(schema.json())\n",
    "#     scala_options = sc._jvm.PythonUtils.toScalaMap(options)\n",
    "#     jc = sc._jvm.com.databricks.spark.xml.functions.from_xml(col._jc if isinstance(col, Column) else col, scala_datatype, scala_options)\n",
    "#     return Column(jc)\n",
    "\n",
    "# new_df = df.select(from_xml(df['body'], schema, {\"rowTag\":\"article\"}))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 81,
   "metadata": {},
   "outputs": [],
   "source": [
    "df = spark.read \\\n",
    "    .format('com.databricks.spark.xml') \\\n",
    "    .options(rowTag='record')\\\n",
    "    .load(file_path)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 149,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "root\n",
      " |-- _abstract-type: string (nullable = true)\n",
      " |-- _id: string (nullable = true)\n",
      " |-- ext-link: struct (nullable = true)\n",
      " |    |-- _VALUE: string (nullable = true)\n",
      " |    |-- _ext-link-type: string (nullable = true)\n",
      " |    |-- _xlink:href: string (nullable = true)\n",
      " |-- fig: struct (nullable = true)\n",
      " |    |-- _id: string (nullable = true)\n",
      " |    |-- _position: string (nullable = true)\n",
      " |    |-- graphic: struct (nullable = true)\n",
      " |    |    |-- _VALUE: string (nullable = true)\n",
      " |    |    |-- _xlink:href: string (nullable = true)\n",
      " |-- italic: array (nullable = true)\n",
      " |    |-- element: string (containsNull = true)\n",
      " |-- p: array (nullable = true)\n",
      " |    |-- element: struct (containsNull = true)\n",
      " |    |    |-- _VALUE: string (nullable = true)\n",
      " |    |    |-- _id: string (nullable = true)\n",
      " |-- sec: array (nullable = true)\n",
      " |    |-- element: struct (containsNull = true)\n",
      " |    |    |-- ext-link: struct (nullable = true)\n",
      " |    |    |    |-- _VALUE: string (nullable = true)\n",
      " |    |    |    |-- _ext-link-type: string (nullable = true)\n",
      " |    |    |    |-- _xlink:href: string (nullable = true)\n",
      " |    |    |-- italic: array (nullable = true)\n",
      " |    |    |    |-- element: string (containsNull = true)\n",
      " |    |    |-- p: array (nullable = true)\n",
      " |    |    |    |-- element: string (containsNull = true)\n",
      " |    |    |-- sub: string (nullable = true)\n",
      " |    |    |-- sup: array (nullable = true)\n",
      " |    |    |    |-- element: string (containsNull = true)\n",
      " |    |    |-- title: string (nullable = true)\n",
      " |-- sub: array (nullable = true)\n",
      " |    |-- element: string (containsNull = true)\n",
      " |-- sup: array (nullable = true)\n",
      " |    |-- element: string (containsNull = true)\n",
      " |-- title: string (nullable = true)\n",
      " |-- xref: struct (nullable = true)\n",
      " |    |-- _VALUE: string (nullable = true)\n",
      " |    |-- _ref-type: string (nullable = true)\n",
      " |    |-- _rid: string (nullable = true)\n",
      "\n"
     ]
    }
   ],
   "source": [
    "df = spark.read \\\n",
    "    .format('com.databricks.spark.xml') \\\n",
    "    .options(rowTag='record')\\\n",
    "    .options(rowTag='abstract')\\\n",
    "    .load(file_path)\n",
    "df.printSchema()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 147,
   "metadata": {},
   "outputs": [],
   "source": [
    "def parse_paragraphs_list(list_of_paragraphs):\n",
    "    if list_of_paragraphs != None:\n",
    "        list_of_paragraphs = [paragraph._VALUE for paragraph in list_of_paragraphs if paragraph._VALUE != None]\n",
    "    else:\n",
    "        list_of_paragraphs = []\n",
    "    return list_of_paragraphs\n",
    "\n",
    "def parse_sections_list(list_of_sections):\n",
    "    if list_of_sections != None:\n",
    "        list_of_sections = [parse_paragraphs_list(section.p) for section in list_of_sections if section.p != None]\n",
    "    else:\n",
    "        list_of_sections = []\n",
    "    return list_of_sections"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 148,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "[[['The sagittal diameter of the cervical spinal canal is important, as a small canal diameter is associated with cervical myelopathy and with a high risk of spinal cord injury after trauma.']],\n",
       " [['Pathological gambling (PG) is a psychiatric disorder characterized by a pre-occupation with thoughts of gambling, repeated attempts to reduce or quit, debt and/or illegal activity, and disruption of personal relationships and/or employment. PG has been estimated to affect between 0.2 and 5.3% of the adult population worldwide (']],\n",
       " [['The sex-steroid hormone estradiol (E2) enhances the psychoactive effects of cocaine, as evidenced by clinical and preclinical studies (']],\n",
       " [['ADCC in HIV-1 has been studied for over 20\\xa0years (']],\n",
       " [['Brain white matter hyperintensities (WMH) are common in community-dwelling older people and are associated with cognitive decline, dementia, stroke, and death (']],\n",
       " [['Spontaneous cerebrospinal fluid (CSF) fistula of the skull base is an uncommon condition associated with multiple etiologies and can have a variable presentation. Of note, obese, middle-aged, and multiparous women most commonly develop this condition.']],\n",
       " [['Since it was first used in 1967, spinal cord stimulation (SCS) has become a commonly used procedure for the management of intractable pain arising from such conditions as failed back surgery syndrome, complex regional pain syndrome, refractory angina, phantom limb pain, postthoracotomy pain syndrome, and reflex sympathetic dystrophy. As with any implanted device, complications would be anticipated. Some of the more common complications include electrode migration, hardware malfunction, cerebrospinal fluid (CSF) leakage, infection, and hematoma. Nerve root/spinal cord injury, allergic reactions, and skin erosions have also been reported but are less frequently encountered.']],\n",
       " [[\"The present study is a cross-sectional observation study. Following metrics were analyzed from the journal's homepage (\"]],\n",
       " [['For the incidence of sacral dysplasia, the upper sacral segment morphology was classified as normal, transitional, and dysplastic according to the radiological characteristics, as defined by Routt et al.']],\n",
       " [['A retrospective chart review was performed to determine outcomes of nonagenarian patients (over 90 years of age) who underwent operative hip fracture repair between January 1st, 2009 and August 21, 2013 at a level 1 trauma center. Institutional Review Board approval was obtained before chart review began. Patients who were treated conservatively, underwent a girdlestone procedure, or resided in a different state than the institution were excluded. This led to a total of 39 patients with a mean age of 93.3 ± 2.3 years being included in this study (34 females, 5 males) ('],\n",
       "  ['All analyses were performed using the IBM SPSS ver. 21.0 (IBM Co., Armonk, NY, USA). Survival was analyzed by the Cox proportional hazards regression model. The chi-square test was used for the analysis of nominal variables. Nonparametric continuous variables were tested by the Mann-Whitney ']],\n",
       " [['From January 1996 to January 2013, 853 patients were admitted to the Inha University Hospital with a femoral neck fracture. Based on the anatomic location, 214 patients were classified as having subcapital fracture. Of the 102 patients who underwent internal fixation using cannulated screws, 97 patients had a follow-up for more than 2 years. The cases excluded from the study were two cases of previous hip fracture, three cases with a history of cancer, eight cases with special medical conditions (one case of vasculopathy, two cases of pathological fractures, two cases of rheumatic arthritis, one case of osteomyelitis, two cases in chronic alcoholics) that may cause AVN. Finally, 84 patients were enrolled as subjects. Diagnosis of subcapital femoral neck fracture was performed by observing the fracture line at the junction between the femoral head and neck. We excluded lower transcervical femoral neck fractures from this study. Their ages when they first sustained fractures, fracture type graded by the Garden classification, the time elapsed before surgery, the degree of reduction after surgery, firm fixation after surgery, and the presence of osteoporosis were assessed. The Garden alignment index was used to assess the levels of reduction measured by plain radiography, which provided the anteroposterior (AP) and lateral views of hip joints immediately after surgery. When the Garden alignment index fell within the range of 160°–180° when seen on AP views and the range of 170°–190° when seen on lateral views, both images were declared as \"excellent,\" as \"good\" if either of the two images fell within the range, and as \"poor\" if neither of the two images fell within the range. Two indices were used to assess firm fixation. As the first index, the average distance between the inserted fixation screws and the subchondral boundaries of the femoral head was measured using the AP and lateral views obtained by plain radiography. As the second index, whether or not the technique of 3-point fixation through the inferomedial cortical bone of the femoral neck was applied when seen on the AP and lateral views obtained by plain radiography. The cannulated screws were placed according to the \\'3 point principle.\\' Both ends of the screws were anchored in the firm bone (calcar femorale), the lateral femoral cortex and the subchondral bone of the femoral head, respectively.']],\n",
       " [['We studied records of 214 TKA patients operated between October 2011 and April 2012. Of these, 120 patients with estimated preoperative vitamin D levels were included in the study. Serum vitamin D levels were checked in an early morning (overnight fasting) blood sample in all patients. All patients had primary osteoarthritis (OA) knee and had undergone unilateral TKA. We divided the 120 patients into two groups based on their preoperative serum vitamin D levels (25-hydroxy vitamin D levels), i.e., group 1 (deficiency group) with vitamin D levels < 30 ng/mL and group 2 (sufficiency group) with vitamin D levels ≥ 30 ng/mL. Accordingly, group 1 comprised 64 patients and group 2, 56 patients. Preoperatively, none of the patients had received vitamin D supplementation. All patients were supplemented with oral vitamin D (0.5 µg/day) from the 14th postoperative day for a duration of 4 weeks.',\n",
       "   'The same senior surgeon (RNM) performed all surgeries using computer navigation (Ci System; Johnson & Johnson, New Brunswick, NJ, USA). Skin incisions were midline, and midvastus arthrotomy was done for all patients. Posterior cruciate was sacrificed and PFC sigma series (Depuy Synthes, Warsaw, IN, USA) posterior stabilized implant was used in all cases. All patients also had the patellar component replaced using a 3-lug, all-polyethylene component. Cement was used for implantation in all cases. Patients were mobilized with a frame from the day after surgery and were generally discharged from the hospital on the fifth postoperative day. After discharge, patients were called for follow-up at 2 weeks, 6 weeks, and 3 months. We recorded body mass index (BMI) of all patients preoperatively. At 3 months, clinical data was collected for Western Ontario and McMaster Osteoarthritis Index (WOMAC), for Short-Form Health Survey (SF-12) questionnaire with Physical Composite Summary (PCS) and Mental Composite Summary (MCS), and for Knee Society Score (KSS) with Knee Score (KS) and Functional Score (FS). Radiographs were obtained pre- and postoperatively while the patient was in hospital, and subsequently at 3 months.',\n",
       "   'Statistical analysis of outcome measures between groups 1 and 2 was with two-sample ']],\n",
       " [['This prospective cohort study included 218 cases that underwent cruciate ligament-substituting total knee arthroplasty from February 2014 to January 2015; a single surgeon performed the operation using the same method without patellar resurfacing in all patients. The lateral patellofemoral ligament and the retinaculum were released, and extensive osteophytes were removed from the patella bone. During surgery, a goniometer was used after insertion of a trial bearing to determine the angle at which impingement occurs between the bearing post and the patella bone when the knee was moved from 90° flexion to full flexion. In this study, impingement was considered positive when contact between the patella bone and the bearing post occurred at knee flexion of 120° or forward displacement of the patella occurred owing to impingement. The study included 139 female and 13 male patients with an average age of 70.6 years (range, 54 to 89 years).',\n",
       "   'The patients were divided into the Biomet mobile type-bearing group (rotating platform knee) and the Corentec fixed-type bearing group (Lospa; Corentec, Seoul, Korea) to compare the impingement frequency; statistical significance was according to implant design.',\n",
       "   'Changes to the joint line were measured at pre- and postoperative standing knee anteroposterior (AP) radiographic views. The most proximal point of the fibula head was selected as the reference point for this study because fibula head enables easy viewing, and is used by many authors. ']],\n",
       " [['For this case-control study, we first obtained local Institutional Review Board approval (identification number 921893). The HADS questionnaire was used to assess associated emotional disorders. The questionnaire contains 14 items yielding ordinal data. Odd items are for the assessment of anxiety whereas even items are for depression status, and each item is scored from 0 to 3; thus, a person may score from 0 to 21 for anxiety or depression (0–7, normal; 8–10, borderline; 11–21, abnormal).']],\n",
       " [['Range of external rotation in 0° and in 90° of abduction was measured using a goniometer. Internal rotation at the back was assessed with the patients sitting upright and their hands placed at their backs as high as possible. Anatomic reference points were the inferior border of the scapula for the seventh thoracic vertebra (T7) and the highest point of the iliac crest for the fourth lumbar vertebra (L4). For ease of statistical analysis, the vertebral levels were converted into numeric values as follows: T1-12 to 1-12, L1-5 to 13-17, and sacrum to 18.',\n",
       "   'Functional outcomes were evaluated using the Rowe score, wherein lesser scores denote worse shoulder condition. The visual analogue scale (VAS) for pain was used to grade the pain suffered by the patient, with 0 being no pain and 10 being the worst pain experienced. Patient satisfaction with regard to the surgical outcome and improvement was also rated using the VAS, with 10 being the best score.',\n",
       "   'The ROM and functional outcome measurements were evaluated during preoperative admission, 6 months and 1 year postoperatively, and annually thereafter. The scores and ranges of motion at the final follow-up visit were used for the statistical analysis.'],\n",
       "  [\"All patients underwent arthroscopic surgery performed by the senior surgeon. Anterior shoulder stabilization consisted of a Bankart or ALPSA lesion fixation with capsular plication, with or without a rotator interval (RI) closure. Additional RI closure using additional suture anchor was confined to patients who demonstrated definite sulcus sign with shoulder in the external rotation position. A concurrent SLAP lesion was repaired between the 10- to 12-o'clock positions on a right shoulder, to avoid anterior labral and capsular plication at the 12- to 2-o'clock positions.\",\n",
       "   'We used the lateral decubitus position for the shoulder arthroscopy with the operated arm pulled by a traction device (ACUFEX; Smith & Nephew, Andover, MA, USA) with approximately 4 kg of weight (10 lb). During anterior capsulolabral reconstruction, an additional lateral traction device was applied for lateral pulling of the arm to obtain easy access to the anterior compartment of the glenohumeral joint. Diagnostic arthroscopy was performed throughout the standard posterior portal as the viewing portal, and an anteroinferior portal as the primary working portal. After overall examination of the glenohumeral joint, the additional anterosuperior portal was used as the secondary working portal for procedures on the anterior labrum or capsule.',\n",
       "   \"For Bankart repair or anterior capsular shift, the abnormally attached labrum or anteroinferior glenohumeral ligament was mobilized from the glenoid neck, then 2- to 3-mm wide subchondral bone was exposed using a motorized burr for the recipient bed. The capsule together with the anterior labrum, when present, were sutured and tacked down using 3 or 4 evenly spaced suture anchors (BIOKNOTLESS; Depuy Mitek Inc., Raynham, MA, USA) between the 2- to 6-o'clock positions, on the right shoulder.\",\n",
       "   \"For repair of a SLAP lesion, we used the trans-rotator cuff portal, as described by O'Brien et al.\"]],\n",
       " [['A retrospective cohort study involving a tertiary referral center was approved by the Chonnam National University Hospital Institutional Review Board to evaluate 70 patients who underwent open reduction and internal fixation with bone graft for established scaphoid nonunion between January 2003 and December 2011.',\n",
       "   'Preoperative two-dimensional radiographic films and CT of the wrist were reviewed for each patient to assess the lunate morphology, define the radiolunate angle and scaphoid fracture location, as well as to measure the capitate-triquetrum (C-T) distance. Only radiographs showing adequate distal radioulnar overlap and a radiusthird metacarpal angle of < 20° from neutral were accepted. An orthopedic surgeon, who had not been involved in treatment, conducted the radiological assessment.',\n",
       "   'Lunate morphology was determined by examining standard posteroanterior (PA) wrist radiographs and CT. If a medial lunate facet could be identified on one of the radiographs and CT, the lunate was classified as type II (']],\n",
       " [[\"Forty healthy volunteers (20 males and 20 females) with a median age of 31 years (range, 21 to 49 years) were recruited for the study. The exclusion criteria were patients with a history of trauma to the hand and wrist and ailments such as arthritis, flexor tenosynovitis, trigger finger, and Dupuytren's disease in the hand and wrist that could limit the range of motion of the fingers.\",\n",
       "   'A transverse line that forms on the palm during MCPJ flexion between the radial end of the thenar crease and the ulnar end of the distal transverse palmar crease (transverse skin folding line) was drawn. Location A was designated at the center of this line that is usually located between the distal and proximal transverse palmar creases. Location B was determined at 1 cm proximal to location A. Location C was determined at 1 cm distal to location A, which is usually located between the distal transverse palmar crease and the proximal digital crease. We applied a volar short arm splint on the dominant hand of the study subjects at each location (']],\n",
       " [['The pilot study and meta-analysis were approved by Institutional Review Board. We collected clinical data for pilot study in spastic diplegia patients with gross motor function classification system level 1 or 2 from March 2010 to March 2013. Patients who underwent only soft tissue surgeries and gait analysis at least 1 year after surgery were included in the study. Patients were excluded for the study if they had hip dislocation or subluxation, had botulinum injection 6 months prior to the study, and had a history of other surgical procedures for cerebral palsy including selective posterior rhizotomy and osteotomies on the femur or tibia. A total of 12 subjects were selected (10 boys and 2 girls). The mean age of the subjects at the time of operation was 10 years (range, 6 years and 3 months to 15 years and 4 months).',\n",
       "   'For meta-analysis, we conducted searches of the literature in EMBASE and PubMed published between March 1980 and March 2014 using the keywords of \"cerebral palsy,\" \"rotation,\" and \"surgery.\" The search results returned 314 articles, and 12 studies matched our inclusion criteria (']],\n",
       " [['Body composition and BMD were measured by wholebody DXA. The accuracy error for BMD measurements was determined according to the standardized protocols from the International Society for Clinical Densitometry.']],\n",
       " [['In Korea University Anam Hospital, neonatal hip screening using selective ultrasound had been performed since 1996 according to the same protocol. All the babies in the neonatal units received a physical examination within 1 week after birth or admission to the neonatal unit by one experienced pediatric orthopedic surgeon. Babies referred from other institutes under the presumptive impression of hip dysplasia also received the same hip screening procedure at the time of visiting the pediatric orthopedic department. From January 2004 to June 2009, 2,686 babies in neonatal units and 43 babies who were referred before the age of 3 months were included in this study. Of the 2,686 babies who were examined in the neonatal unit, 1,453 babies (54%) were boys and the remaining 1,233 babies (46%) were girls. The distribution of body weight at birth was as follows: 67 babies (2.5%), under 1 kg; 107 (4.0%), 1–1.5 kg; 496 (18.5%), 1.5–2.5 kg; 1,921 (71.5%), 2.5–4.0 kg; and 95 (3.5%), over 4.0 kg.',\n",
       "   'The Barlow and Ortolani test was used for the physical examinations. Based on the examination results, hips were subjectively divided into the following 5 groups: (1) normal; (2) noisy hip or suspicious instability; (3) subluxatable; (4) dislocatable or dislocated; and (5) limited abduction similar to the clinical assessment of hip stability described by Laborie et al.']],\n",
       " [['In February 2010, a 39-year-old man fractured both bones of the forearm after his arm was caught in a conveyor belt; he underwent a surgery at another hospital. An intramedullary forearm rod (Acumed, Hillsboro, OR, USA) was used in the ulna ('],\n",
       "  ['We recommend 2 skin incisions to minimize the total incision length. The distal incision is approximately the size of 2 inter-interlocking holes, and the proximal incision is about 1 cm, big enough to pass the nail (']],\n",
       " [['This case report was approved by the Institutional Review Board of Tokoin Teaching Hospital and the requirement for obtaining informed consent was waived. This case is of a 23-year-old young man who was brought to the surgical emergency department by firefighters after about 3 hours, in January 2013 when he was hit by a truck that probably dragged him some distance away. On admission, he showed an altered conscious state (Glasgow Coma Scale, 9) without focal signs and hemodynamic shock; on the left side, an inguinal wound about 8 cm through which the femoral head was visible in the anterior and upper position; the hip in extension, abduction and external rotation and shortening of 3 cm with no signs of neurovascular complication downstream. We also noted a perineal wound and skin abrasions adjacent to the iliac crest on the right side (']],\n",
       " [['A 71-year-old woman presented with progressive complaints in the right knee. Complaints at presentation were: daily morning stiffness, starting pain, giving way, and pain at night, causing impairment in daily activities and decrease in the walking distance. Conservative therapy (medication and application of a valgus knee brace) was not sufficient. Physical examination showed a clinical varus leg axis, a stable medial collateral ligament, 15° of impaired extension, and 100° of flexion of the right knee. A significant swelling was noted on the dorsal side of the knee. X-ray showed significant osteoarthritis with marked medial joint space narrowing and substantial SO with multiple round calcifications in the intra-articular joint space and in the suprapatellar bursa (']],\n",
       " [['A 25-year-old male patient was admitted to the hospital. He suffered from trauma in the right hand in a motorcycle accident. He complained of swelling, functional limitation, and pain in the injured hand. Physical examination showed dorsal deformity of the metacarpal-trapezoid joint without distal neurovascular deficit. Radiographs showed metacarpal-trapezoid joint dislocation without fracture (']],\n",
       " [['The first case was a 29-year-old pregnant woman presenting with a DRF. At the follow-up X-ray after closed reduction, an increase in the dorsal tilt of the radius and flexion of the lunate was observed; however, the operation was postponed because the patient was pregnant. At 8 months after the trauma, the patient complained of painful limited range of motion in the wrist. The preoperative dorsal tilt was 27° and the contralateral volar tilt was 7°. Corrective dorsal open wedge osteotomy of the distal radius was performed ']],\n",
       " [['It is widely assumed that Low Back Pain (LBP) with mechanical pain behavior is significantly influenced by biomechanical factors [']],\n",
       " [['Severe infections are a leading cause of the observed high morbidity and mortality in intensive care unit (ICU) patients. A key element for optimal patient outcomes includes sufficiently high antimicrobial concentration levels [']],\n",
       " [['Taiwan is an island country populated by approximately 23 million people with a gross domestic product per capita of $22,632 in 2014 [']],\n",
       " [['There is a significant difference in the prevalence of atopic conditions between Finnish and Russian Karelia although these two areas are geographically adjacent and inhabitants of these two areas share partly the same ancestry. Living conditions in Russian Karelia are rural and similar to those in Finland in the 1950\\xa0s, and the socio-economic gap between the two areas is prominent [']],\n",
       " [['Childhood undernutrition in the form of chronic and acute malnutrition coexist in many populations in developing countries. Unlike acute malnutrition which is associated with immediate crisis such as periodic food shortages, chronic malnutrition is due to inadequate nutrition over a prolonged period resulting from latent poverty, chronic food insecurity, poor feeding practices and repeated episodes of health problems (such as infections) or poor health services in an unhealthy environment [']],\n",
       " [['The Simple Knowledge Organization System (SKOS) standard was introduced to the wider community by a 2005 World Wide Web Consortium (W3C) working draft [']],\n",
       " [['Hereditary cerebellar ataxias are neurodegenerative diseases usually having a detrimental impact on patients’ quality of life or even leading to premature death. Patients suffer from a cerebellar syndrome, also combined in most of the hereditary ataxias with extracerebellar neurological symptoms, and in some cases other organ systems are also affected (for review see [']],\n",
       " [['In animals, miRNAs interact with their targets through imperfect base pairing. The limited sequence interactions required by miRNAs to direct regulation allows a single miRNA to potentially regulate hundreds of targets in multiple pathways. Although miRNAs are flexible in their targeting ability, a large body of work has proposed a series of rules that predict canonical miRNA targeting [']],\n",
       " [['Chronic obstructive pulmonary disease (COPD) is characterized by an enhanced pulmonary and systemic inflammatory response to inhaled particles and gases, particularly those found in tobacco smoke. COPD patients have an increased risk of developing non-small cell lung cancer (NSCLC) that is independent of smoking pack-year history [']],\n",
       " [['Typhoid fever is a major public health concern in developing countries. It accounts for an estimated 22 million cases and approximately 1,100,000 deaths annually worldwide [']],\n",
       " [['Duchenne muscular dystrophy (DMD) affects approximately 1 in 5000 newborn boys in whom ']],\n",
       " [['Many patients with stage 5 chronic kidney disease that are not treated with renal replacement therapy are often provided by primary care physicians, many independently without consultation from nephrology. Providing conservative care in the primary care setting is complex and incompletely understood.'],\n",
       "  ['This paper presents a mixed methods study to comprehensively understand barriers and facilitators to community-based kidney conservative management by primary care physicians, which can serve as an example for other investigators interested in mixed methods research within the field of nephrology.'],\n",
       "  ['There is a large proportion of older adults with stage 5 chronic kidney disease (CKD) not treated with renal replacement therapy. In Alberta, Canada, 51\\xa0% of people with stage 5 CKD not treated with renal replacement therapy are aged 75\\xa0years and older [']],\n",
       " [['Frozen dough technology has been used in bakery industry to provide consumers with high-quality fresh bakery and convenience. However, cellular macromolecules, including proteins, nucleic acids and lipids of the yeast used in frozen dough, could be seriously damaged under the freezing and the subsequent thawing treatments, leading to inhibition of cell growth, cell viability and the leavening ability [']],\n",
       " [['Men who have sex with men (MSM) continue to be the group most affected by HIV infection ['],\n",
       "  ['Public Health England collect annual surveillance relating to diagnosis and routes of transmission; these figures indicate that men of white ethnicity comprise 84\\xa0% (38,429 of 45,679) of cases of newly diagnosed MSM with HIV in E&W with a route of exposure through sex with men [']],\n",
       " [['Out of 56 million deaths in 2012 globally, 38 million were due to NCDs, principally diabetes, cardiovascular diseases, cancer and chronic respiratory diseases [']],\n",
       " [['A large body of research exists which suggests that medical practitioners and medical students may be at increased risk of a number of personal and psychological problems. Dealing with the suffering of others can have a negative impact on health professionals, resulting in psychological illnesses, suicide, depression, anxiety and alcoholism [']],\n",
       " [['Large-scale, functional community health worker programs (CHW-P) make important strides to counteract health workforce shortages in low and middle-income countries (LMICs) by extending primary healthcare to rural and underserved communities [']],\n",
       " [['Health professionals often manage medical problems in critical situations under time pressure and on the basis of vague information. Remarkably, observation of experienced clinicians making medical decisions has revealed how quickly diagnostic and therapeutic decisions are made. This complex process termed “medical problem solving” or “clinical reasoning” has been investigated for more than three decades ['],\n",
       "  ['In surgery, the patient presenting with acute abdominal pain requires clinical assessment and therapy within 2\\xa0h. Distinguishing between various differential diagnoses in a time-efficient way is critical and requires the “fine art of diagnostics” [']],\n",
       " [['Many parallels exist between the metabolic profiles of cancer cells and normal proliferating cells, including the use of aerobic glycolysis, selective expression of metabolic enzymes with distinct regulatory features, and elevation of amino acid consumption and biosynthesis [']],\n",
       " [['Head and neck cancers include malignant neoplasms that arise from many sites within the upper aerodigestive tract, with the most common sites being the oropharynx, hypopharynx, larynx, and oral cavity [']],\n",
       " [['Metabolic syndrome (MetS) is a cluster of factors associated with a higher risk of type 2 diabetes, cardiovascular diseases and mortality [']],\n",
       " [['High rates of childhood obesity are a worldwide concern ['],\n",
       "  ['The objectives of this systematic review are to answer the following research questions:']],\n",
       " [['Breast sarcomas are very uncommon and make up less than 1\\xa0% of all primary breast malignancies. Primary osteosarcoma accounts for about 12.5\\xa0% of all breast sarcomas.',\n",
       "   'The histogenesis of primary osteosarcoma of the breast is unknown, prognosis and optimal treatment remain uncertain because of the rarity of this tumor. Published data consist almost exclusively of single case reports, with one major series [']],\n",
       " [['In response to changing environmental conditions, such as nutrient fluxes or the presence of stress, eukaryotic cells adapt by modulating chromatin structure and their gene expression programs [']],\n",
       " [['Qatar is a country of over two million people located in the Arabian Gulf region, which, together with another five countries, form the Gulf Cooperation Council (GCC). Recent development in social and dietary changes has resulted in lifestyle-related diseases such as diabetes, cardiovascular disease and dental caries [']],\n",
       " [['Papillary Thyroid Carcinoma (PTC) is the most common type of thyroid cancer, accounting for more than 75\\xa0% of all thyroid malignancies. The incidence of PTC is increasing more rapidly than any other cancer [']],\n",
       " [['Results-based financing (RBF) in health can be defined as “a cash payment or non-monetary transfer made to a national or sub-national government, manager, provider, payer or consumer of health services after predefined results have been attained and verified” [']],\n",
       " [['Gliomas are the most common primary neoplasms of the brain [']],\n",
       " [['Alzheimer’s disease (AD) is a progressive neurodegenerative brain disorder that is characterized by two hallmarks: intracellular neurofibrillary tangles and extracellular amyloid-β (Aβ) plaques in the regions of the brain that are responsible for learning and memory. Aβ, peptides of 39–43 amino acids produced by the sequential cleavage of β- and r-secretase at the C terminal of amyloid precursor protein, are believed to play a central role in the neurodegeneration and subsequent cognitive abnormalities of AD brains [']],\n",
       " [['This paper serves to report on a presentation that was made at the International Symposium on Animal Functional Genomics conference in Piacenza, Italy that was held between July 27 and 29, 2015. Consequently, some of the material reported here has been published elsewhere [']],\n",
       " [['The temporal organization of activity/rest or sleep/wake rhythms for mammals is regulated by the interaction of light/dark cycle and circadian clocks. Several lines of evidence indicate that the known master circadian clock, suprachiasmatic (SCN), operates quite similarly in nocturnal and diurnal mammals. The oscillations of clockwork genes, such as Bmal1, Per1, and Per2, are in the same phase in the SCN, regardless whether the mammals are diurnal or nocturnal [']],\n",
       " [],\n",
       " [['CO, cardiac output; IVC, inferior vena cava; LVEDV, left ventricular end diastolic volume; PEEP, positive end-expiratory pressure; PVR, pulmonary vascular resistance; RAP, right atrial pressure; STE, speckle tracking echocardiography; SVR, systemic vascular resistance']],\n",
       " [['Mature cystic teratomas (MCTs; dermoid cysts) occur most commonly in the ovary, with an incidence ranging from 5 to 25\\xa0% of all ovarian neoplasms [']],\n",
       " [['Chronic obstructive pulmonary disease (COPD) represents one of the main causes of morbidity and mortality, and worldwide nearly 3 million people die from COPD every year [']],\n",
       " [['While many children find social interaction stress reducing, children with autism spectrum disorder (ASD) often find social interaction stress-inducing [']],\n",
       " [['In this review of published and unpublished literature, we looked for evidence, including clinical studies (completed, terminated or ongoing), supportive studies (subgroup analyses and models) and commentaries discussing the material used for the EMA approval (letters and systematic or non-systematic reviews) about the efficacy, effectiveness or efficiency of nalmefene, via a search of PubMed, clinicaltrials.gov, isrctn.com, clinicaltrialsregister.eu and lundbeck.com/trials, up to and including 25 April 2016, using the term ‘nalmefene’. In addition, we used approval documents from the EMA, the NICE, the French transparency committee, and a earlier systematic review and meta-analysis by our team [']],\n",
       " [['In neurosurgery, it is imperative to avoid intracranial bleeding. Perioperative bleeding can be associated with a number of factors including antihemostatic drugs and coagulation status but is also linked to the tumor’s vascularity, type, size and localization and the use of local hemostatics (Gerlach et al. ']],\n",
       " [['The prevalence of gestational diabetes mellitus (GDM) ranges between 0.7 and 10.1\\xa0% according to different diagnostic criteria worldwide [']],\n",
       " [['The apelin gene encodes a preproprotein of 77 amino acids that is processed into multiple shorter peptides including apelin-36, apelin-17, apelin-13 and apelin-12 (']],\n",
       " [['Climate change is likely to influence infectious disease dynamics, with many pathogens, especially those with complex life cycles or those infecting ectothermic hosts, predicted to change in severity or range as the earth continues to warm [']],\n",
       " [['Fungal pathogens are the most damaging disease agents in global crop production. Despite chemical controls and plant resistance breeding, around one-quarter of global production is lost, enough to feed hundreds of millions [']],\n",
       " [['Emerging fungal infections continue to pose significant threats to animals, plants and humans, causing disease outbreaks and even posing a threat to food security [']],\n",
       " [['Fungi represent a major threat to human health accounting collectively for more than a billion skin infections, more than 100 million mucosal infections, 10 million serious allergies and more than a million deaths each year. Global mortality owing to fungal infections is greater than for malaria and breast cancer and is equivalent to that for tuberculosis (TB) and HIV [']],\n",
       " [['Too many people die from AIDS, most of them adults in the prime of life. The majority of these deaths are attributable to opportunistic infections. As an example, across the world, HIV infection is second only to traffic accidents as a cause of death in adolescents [']],\n",
       " [['Emerging fungal pathogens and fungal-like organisms are increasingly threatening natural populations of animals and plants [']],\n",
       " [],\n",
       " [['Iodine deficiency disorders remain a public health target worldwide, and 35% of the world’s population is estimated to have insufficient iodine intake.']],\n",
       " [['Population aging is a global phenomenon. Many countries are implementing wide-ranging healthcare and long-term care (LTC) reforms to maintain sustainability in their financing and care provision systems. Reducing caregiver burden is particularly important, as caregiver burden hinders the health outcomes of disabled older adults and caregiver performance. Family caregivers experience a substantial sense of burden,']],\n",
       " [['The purposes of the present study were to conduct a long-term (20-year) prospective investigation of the relationship of the combination of ']],\n",
       " [['In China, cardiovascular disease (CVD) has become a leading cause of death over the past two decades as a result of economic growth and associated changes in lifestyle and diet, including unhealthy dietary patterns and an increase in the prevalence of overweight.']],\n",
       " [['Lung cancer is one of the leading causes of death from cancer for men and women in Japan.']],\n",
       " [['Hypertension is considered to be the leading risk factor for cardiovascular mortality and accounts for a large proportion of premature deaths in China as well as in Western countries.']],\n",
       " [],\n",
       " [['Cell stage–specific kinases are important for the progression of mitosis. These kinases play a role in specific pathways to ensure that chromosomes segregate properly to daughter cells to prevent aneuploidy. Among the mitotic kinases, Aurora B plays a central role in the maintenance of genome stability by activating the spindle assembly checkpoint in response to improper kinetochore–microtubule attachment and tension during early mitosis (']],\n",
       " [['For chromosomes to segregate equally during mitotic cell division, they must be bioriented on the spindle. This requires that the duplicated sister chromatids remain paired until anaphase. One mechanism of chromatid pairing is provided by the entanglement of the newly replicated sister DNA molecules (']],\n",
       " [['Lysosomes are heterogenous organelles that are able to fuse with the plasma membrane (PM; ']],\n",
       " [['The elongated nature of skeletal myofibers combined with the forces of muscle contraction renders the muscle plasma membrane prone to small lesions and tears (']],\n",
       " [['In most proliferative cells, the centrosome acts as the primary microtubule-organizing center (MTOC). Although it has been long appreciated that differentiation induces formation of noncentrosomal microtubule (MT) arrays in many tissues and cell types, including epithelium, neurons, and muscle, the mechanisms controlling inactivation of the centrosome during this process remain poorly characterized (']],\n",
       " [['Proteins of the secretory pathway are often modified after translocation across or insertion into the membrane of the ER (']],\n",
       " [['Inflammasomes are multimeric complexes formed in response to a variety of physiological and pathogenic stimuli. Inflammasome activation is an essential component of the innate immune response and is critical for the clearance of pathogens or damaged cells. However, overt inflammasome activation is also a major driver of autoimmune and metabolic disorders, underlying the importance of understanding this process in physiological and pathological contexts.',\n",
       "   'The inflammasome sensors are grouped according to their structural features into nucleotide-binding domain–like receptors (NLRs), absent in melanoma 2–like receptors (ALRs), and the recently identified pyrin. These receptors have the ability to assemble inflammasomes and activate the cysteine protease caspase-1. In addition to the sensor (NLR, ALR, or pyrin) and enzymatic component (caspase-1), most inflammasomes also use an adaptor molecule known as ASC (apoptosis-associated speck-like protein containing a caspase activation and recruitment domain). Upon detecting specific stimuli, the activated sensor nucleates ASC to form a discrete foci, or “speck,” within the activated cell. Nucleated ASC sequentially recruits caspase-1, which undergoes proximity induced autocatalytic cleavage to produce the active subunits p10 and p20. These active caspase-1 subunits can then proteolytically process cytokines, IL-1β and IL-18, and induce a specific form of inflammatory cell death termed pyroptosis. Activated caspase-1 thus provides the host cell with dual defense mechanisms through the release of mature cytokines and removal of the infected or damaged cell. Inflammasome assembly is thus a coordinated signaling event essential for mounting an appropriate immune response after pathogenic or sterile insults. Although most studies have used in vitro culture systems and reconstitution assays to analyze inflammasome activation, recent advances have allowed visualization of these processes in vivo after an infectious insult at the level of a single cell (']],\n",
       " [['In cytokinesis, the final stage of cell division, an actomyosin-based contractile ring physically divides the cell into two genetically equivalent daughter cells. Our understanding of cytokinesis has been greatly influenced by classical experiments in which spindles and/or cells were repositioned or micromanipulated. These perturbations demonstrated that the spindle induces furrow formation during a specific time interval after anaphase onset (']],\n",
       " [],\n",
       " [],\n",
       " [],\n",
       " [],\n",
       " [],\n",
       " [],\n",
       " [],\n",
       " [['Acute Myeloid Leukemia (AML) is a predominantly fatal hematopoietic malignancy']],\n",
       " [['Mammalian sirtuin 6 (SIRT6) belongs to the Sir2 (silencing information regulator 2) family of enzymes, which were initially reported as nicotinamide adenine dinucleotide (NAD']],\n",
       " [['Cytokines and chemokines recruit and activate specialized effector cells to sites of inflammation']],\n",
       " [['For measurement of intra-subunit conformational changes, i.e. lid-closure and N/M-domain association, we formed hetero-dimers consisting of one wild-type and one PET reporter-containing subunit, thus avoiding complications in data interpretation that would arise from two fluorescence probes located on the same Hsp90 dimer. For measurement of β-strand swap, we formed hetero-dimers consisting of one subunit that contained fluorescently modified A2C and one subunit that contained the E162W mutation. We triggered closure of the molecular clamp by rapidly adding excess of the non-hydrolysable ATP-analogue adenosine 5′-[β,γ-imido]triphosphate (AMP-PNP) to reporter-containing Hsp90 samples and measured the time-dependence of fluorescence intensities in solution. Nucleotide binds to both subunits of Hsp90, which is the physiologically active state']],\n",
       " [],\n",
       " [['For each trial, the rewarded object position (left versus right) within a particular context was randomized. Each trial was randomly assigned to one of the two contexts, with constraints predetermining that 20% of all trials would repeat contextual conditions.'],\n",
       "  ['Local field potentials (LFPs) were recorded at 1000 Hz, while spiking activity (for phase-spike and selectivity index analyses) was captured at 400–8000 Hz and digitized to 40 kHz using a Multichannel Acquisition Processor (Plexon). Single clusters were manually isolated using Offline Sorter (Plexon) waveform dimensions (relative amplitudes across each wire, peak to valley difference, and waveform width). Identified units were screened for inter-spike intervals found to be shorter than the neuronal refractory period (< 2 ms).'],\n",
       "  ['Electrode tip lesions were generated by passing 40 µA current through each recording wire for 30 s or until the connection was severed. Rats then received an overdose of Euthasol and were perfused intracardially with 10% formalin phosphate following 0.9% saline. Brains were extracted and left in 30% sucrose solution for 48 hrs. Frozen coronal sections (40 µm) were collected from each relevant region and stained with cresyl violet. Electrode tip lesion locations were confirmed within mPFC (PL/ IL), dHPC, or vHPC']],\n",
       " [],\n",
       " [],\n",
       " [['Flavivirus is a genus of viruses belonging to the family ']],\n",
       " []]"
      ]
     },
     "execution_count": 148,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df.select(\"sec\")\\\n",
    "  .rdd\\\n",
    "  .map(lambda row: parse_sections_list(row['sec']))\\\n",
    "  .collect()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 82,
   "metadata": {
    "scrolled": false
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>abstract</th>\n",
       "      <th>body</th>\n",
       "      <th>custom-meta-group</th>\n",
       "      <th>elocation-id</th>\n",
       "      <th>ext-link</th>\n",
       "      <th>fig</th>\n",
       "      <th>funding-group</th>\n",
       "      <th>graphic</th>\n",
       "      <th>header</th>\n",
       "      <th>history</th>\n",
       "      <th>...</th>\n",
       "      <th>permissions</th>\n",
       "      <th>pub-date</th>\n",
       "      <th>sec</th>\n",
       "      <th>self-uri</th>\n",
       "      <th>sub</th>\n",
       "      <th>sup</th>\n",
       "      <th>td</th>\n",
       "      <th>tr</th>\n",
       "      <th>volume</th>\n",
       "      <th>xref</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>NaN</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>(2016-06-01, oai:pubmedcentral.nih.gov:4868590...</td>\n",
       "      <td>None</td>\n",
       "      <td>...</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>NaN</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>None</td>\n",
       "      <td>([et al], None, (None, None, [et al], None, (P...</td>\n",
       "      <td>None</td>\n",
       "      <td>NaN</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>(2016-06-01, oai:pubmedcentral.nih.gov:4869041...</td>\n",
       "      <td>None</td>\n",
       "      <td>...</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>NaN</td>\n",
       "      <td>[(Verdejo-Garcia , bibr, bib46)]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>NaN</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>(2016-06-01, oai:pubmedcentral.nih.gov:4869059...</td>\n",
       "      <td>None</td>\n",
       "      <td>...</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>NaN</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>NaN</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>(2016-06-01, oai:pubmedcentral.nih.gov:4869150...</td>\n",
       "      <td>None</td>\n",
       "      <td>...</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>[(s0020, None, None, [htfr], 2.2, [EGFP-CEM-NK...</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>NaN</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>NaN</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>(2016-06-01, oai:pubmedcentral.nih.gov:4869590...</td>\n",
       "      <td>None</td>\n",
       "      <td>...</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>NaN</td>\n",
       "      <td>[(Debette et al., 2011, bibr, bib10), (van Dij...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>98</th>\n",
       "      <td>None</td>\n",
       "      <td>(None, [(The hippocampus is critical to memory...</td>\n",
       "      <td>None</td>\n",
       "      <td>NaN</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>(2016-12-20, oai:pubmedcentral.nih.gov:4961615...</td>\n",
       "      <td>None</td>\n",
       "      <td>...</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>[&lt;xref&gt;12&lt;/xref&gt;]</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>NaN</td>\n",
       "      <td>[(Fig. 1, fig, F1), (Supplementary Fig. 1, sup...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>99</th>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>NaN</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>(2016-12-20, oai:pubmedcentral.nih.gov:4964931...</td>\n",
       "      <td>None</td>\n",
       "      <td>...</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>NaN</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>100</th>\n",
       "      <td>None</td>\n",
       "      <td>([BRCA2, BRCA2], [(Mutations in the breast can...</td>\n",
       "      <td>None</td>\n",
       "      <td>NaN</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>(2016-12-20, oai:pubmedcentral.nih.gov:4973888...</td>\n",
       "      <td>None</td>\n",
       "      <td>...</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>NaN</td>\n",
       "      <td>[(5, bibr, R5), (7, bibr, R7), (8, bibr, R8)]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>101</th>\n",
       "      <td>None</td>\n",
       "      <td>(None, [(Flaviviruses have a single-stranded, ...</td>\n",
       "      <td>None</td>\n",
       "      <td>NaN</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>(2016-12-20, oai:pubmedcentral.nih.gov:5003325...</td>\n",
       "      <td>None</td>\n",
       "      <td>...</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>[&lt;xref&gt;17&lt;/xref&gt;&lt;xref&gt;18&lt;/xref&gt;]</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>NaN</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>102</th>\n",
       "      <td>None</td>\n",
       "      <td>(None, [(To identify proteins essential to a b...</td>\n",
       "      <td>None</td>\n",
       "      <td>NaN</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>(2016-12-20, oai:pubmedcentral.nih.gov:5010022...</td>\n",
       "      <td>None</td>\n",
       "      <td>...</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>[&lt;xref&gt;6&lt;/xref&gt;, &lt;xref&gt;7&lt;/xref&gt;, &lt;xref&gt;8&lt;/xref...</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>NaN</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>103 rows × 26 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "    abstract                                               body  \\\n",
       "0       None                                               None   \n",
       "1       None  ([et al], None, (None, None, [et al], None, (P...   \n",
       "2       None                                               None   \n",
       "3       None                                               None   \n",
       "4       None                                               None   \n",
       "..       ...                                                ...   \n",
       "98      None  (None, [(The hippocampus is critical to memory...   \n",
       "99      None                                               None   \n",
       "100     None  ([BRCA2, BRCA2], [(Mutations in the breast can...   \n",
       "101     None  (None, [(Flaviviruses have a single-stranded, ...   \n",
       "102     None  (None, [(To identify proteins essential to a b...   \n",
       "\n",
       "    custom-meta-group  elocation-id ext-link   fig funding-group graphic  \\\n",
       "0                None           NaN     None  None          None    None   \n",
       "1                None           NaN     None  None          None    None   \n",
       "2                None           NaN     None  None          None    None   \n",
       "3                None           NaN     None  None          None    None   \n",
       "4                None           NaN     None  None          None    None   \n",
       "..                ...           ...      ...   ...           ...     ...   \n",
       "98               None           NaN     None  None          None    None   \n",
       "99               None           NaN     None  None          None    None   \n",
       "100              None           NaN     None  None          None    None   \n",
       "101              None           NaN     None  None          None    None   \n",
       "102              None           NaN     None  None          None    None   \n",
       "\n",
       "                                                header history  ...  \\\n",
       "0    (2016-06-01, oai:pubmedcentral.nih.gov:4868590...    None  ...   \n",
       "1    (2016-06-01, oai:pubmedcentral.nih.gov:4869041...    None  ...   \n",
       "2    (2016-06-01, oai:pubmedcentral.nih.gov:4869059...    None  ...   \n",
       "3    (2016-06-01, oai:pubmedcentral.nih.gov:4869150...    None  ...   \n",
       "4    (2016-06-01, oai:pubmedcentral.nih.gov:4869590...    None  ...   \n",
       "..                                                 ...     ...  ...   \n",
       "98   (2016-12-20, oai:pubmedcentral.nih.gov:4961615...    None  ...   \n",
       "99   (2016-12-20, oai:pubmedcentral.nih.gov:4964931...    None  ...   \n",
       "100  (2016-12-20, oai:pubmedcentral.nih.gov:4973888...    None  ...   \n",
       "101  (2016-12-20, oai:pubmedcentral.nih.gov:5003325...    None  ...   \n",
       "102  (2016-12-20, oai:pubmedcentral.nih.gov:5010022...    None  ...   \n",
       "\n",
       "     permissions pub-date                                                sec  \\\n",
       "0           None     None                                               None   \n",
       "1           None     None                                               None   \n",
       "2           None     None                                               None   \n",
       "3           None     None  [(s0020, None, None, [htfr], 2.2, [EGFP-CEM-NK...   \n",
       "4           None     None                                               None   \n",
       "..           ...      ...                                                ...   \n",
       "98          None     None                                               None   \n",
       "99          None     None                                               None   \n",
       "100         None     None                                               None   \n",
       "101         None     None                                               None   \n",
       "102         None     None                                               None   \n",
       "\n",
       "    self-uri   sub                                                sup    td  \\\n",
       "0       None  None                                               None  None   \n",
       "1       None  None                                               None  None   \n",
       "2       None  None                                               None  None   \n",
       "3       None  None                                               None  None   \n",
       "4       None  None                                               None  None   \n",
       "..       ...   ...                                                ...   ...   \n",
       "98      None  None                                  [<xref>12</xref>]  None   \n",
       "99      None  None                                               None  None   \n",
       "100     None  None                                               None  None   \n",
       "101     None  None                   [<xref>17</xref><xref>18</xref>]  None   \n",
       "102     None  None  [<xref>6</xref>, <xref>7</xref>, <xref>8</xref...  None   \n",
       "\n",
       "       tr volume                                               xref  \n",
       "0    None    NaN                                               None  \n",
       "1    None    NaN                   [(Verdejo-Garcia , bibr, bib46)]  \n",
       "2    None    NaN                                               None  \n",
       "3    None    NaN                                               None  \n",
       "4    None    NaN  [(Debette et al., 2011, bibr, bib10), (van Dij...  \n",
       "..    ...    ...                                                ...  \n",
       "98   None    NaN  [(Fig. 1, fig, F1), (Supplementary Fig. 1, sup...  \n",
       "99   None    NaN                                               None  \n",
       "100  None    NaN      [(5, bibr, R5), (7, bibr, R7), (8, bibr, R8)]  \n",
       "101  None    NaN                                               None  \n",
       "102  None    NaN                                               None  \n",
       "\n",
       "[103 rows x 26 columns]"
      ]
     },
     "execution_count": 82,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df.toPandas()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 89,
   "metadata": {},
   "outputs": [
    {
     "ename": "Py4JJavaError",
     "evalue": "An error occurred while calling z:org.apache.spark.api.python.PythonRDD.runJob.\n: org.apache.spark.SparkException: Job aborted due to stage failure: Task 0 in stage 94.0 failed 1 times, most recent failure: Lost task 0.0 in stage 94.0 (TID 458, madalinlinasmbp, executor driver): org.apache.spark.api.python.PythonException: Traceback (most recent call last):\n  File \"/usr/local/Cellar/apache-spark/3.0.1/libexec/python/lib/pyspark.zip/pyspark/worker.py\", line 605, in main\n    process()\n  File \"/usr/local/Cellar/apache-spark/3.0.1/libexec/python/lib/pyspark.zip/pyspark/worker.py\", line 597, in process\n    serializer.dump_stream(out_iter, outfile)\n  File \"/usr/local/Cellar/apache-spark/3.0.1/libexec/python/lib/pyspark.zip/pyspark/serializers.py\", line 271, in dump_stream\n    vs = list(itertools.islice(iterator, batch))\n  File \"/usr/local/Cellar/apache-spark/3.0.1/libexec/python/pyspark/rdd.py\", line 1440, in takeUpToNumLeft\n    yield next(iterator)\n  File \"/usr/local/Cellar/apache-spark/3.0.1/libexec/python/lib/pyspark.zip/pyspark/util.py\", line 107, in wrapper\n    return f(*args, **kwargs)\n  File \"<ipython-input-89-913fc787ca91>\", line 15, in <lambda>\n  File \"<ipython-input-89-913fc787ca91>\", line 7, in parser_xml\n  File \"/Users/madalina-alinaracovita/opt/anaconda3/lib/python3.8/xml/etree/ElementTree.py\", line 1320, in XML\n    parser.feed(text)\n  File \"<string>\", line None\nxml.etree.ElementTree.ParseError: not well-formed (invalid token): line 107, column 29\n\n\tat org.apache.spark.api.python.BasePythonRunner$ReaderIterator.handlePythonException(PythonRunner.scala:503)\n\tat org.apache.spark.api.python.PythonRunner$$anon$3.read(PythonRunner.scala:638)\n\tat org.apache.spark.api.python.PythonRunner$$anon$3.read(PythonRunner.scala:621)\n\tat org.apache.spark.api.python.BasePythonRunner$ReaderIterator.hasNext(PythonRunner.scala:456)\n\tat org.apache.spark.InterruptibleIterator.hasNext(InterruptibleIterator.scala:37)\n\tat scala.collection.Iterator.foreach(Iterator.scala:941)\n\tat scala.collection.Iterator.foreach$(Iterator.scala:941)\n\tat org.apache.spark.InterruptibleIterator.foreach(InterruptibleIterator.scala:28)\n\tat scala.collection.generic.Growable.$plus$plus$eq(Growable.scala:62)\n\tat scala.collection.generic.Growable.$plus$plus$eq$(Growable.scala:53)\n\tat scala.collection.mutable.ArrayBuffer.$plus$plus$eq(ArrayBuffer.scala:105)\n\tat scala.collection.mutable.ArrayBuffer.$plus$plus$eq(ArrayBuffer.scala:49)\n\tat scala.collection.TraversableOnce.to(TraversableOnce.scala:315)\n\tat scala.collection.TraversableOnce.to$(TraversableOnce.scala:313)\n\tat org.apache.spark.InterruptibleIterator.to(InterruptibleIterator.scala:28)\n\tat scala.collection.TraversableOnce.toBuffer(TraversableOnce.scala:307)\n\tat scala.collection.TraversableOnce.toBuffer$(TraversableOnce.scala:307)\n\tat org.apache.spark.InterruptibleIterator.toBuffer(InterruptibleIterator.scala:28)\n\tat scala.collection.TraversableOnce.toArray(TraversableOnce.scala:294)\n\tat scala.collection.TraversableOnce.toArray$(TraversableOnce.scala:288)\n\tat org.apache.spark.InterruptibleIterator.toArray(InterruptibleIterator.scala:28)\n\tat org.apache.spark.api.python.PythonRDD$.$anonfun$runJob$1(PythonRDD.scala:154)\n\tat org.apache.spark.SparkContext.$anonfun$runJob$5(SparkContext.scala:2139)\n\tat org.apache.spark.scheduler.ResultTask.runTask(ResultTask.scala:90)\n\tat org.apache.spark.scheduler.Task.run(Task.scala:127)\n\tat org.apache.spark.executor.Executor$TaskRunner.$anonfun$run$3(Executor.scala:446)\n\tat org.apache.spark.util.Utils$.tryWithSafeFinally(Utils.scala:1377)\n\tat org.apache.spark.executor.Executor$TaskRunner.run(Executor.scala:449)\n\tat java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1128)\n\tat java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:628)\n\tat java.base/java.lang.Thread.run(Thread.java:834)\n\nDriver stacktrace:\n\tat org.apache.spark.scheduler.DAGScheduler.failJobAndIndependentStages(DAGScheduler.scala:2059)\n\tat org.apache.spark.scheduler.DAGScheduler.$anonfun$abortStage$2(DAGScheduler.scala:2008)\n\tat org.apache.spark.scheduler.DAGScheduler.$anonfun$abortStage$2$adapted(DAGScheduler.scala:2007)\n\tat scala.collection.mutable.ResizableArray.foreach(ResizableArray.scala:62)\n\tat scala.collection.mutable.ResizableArray.foreach$(ResizableArray.scala:55)\n\tat scala.collection.mutable.ArrayBuffer.foreach(ArrayBuffer.scala:49)\n\tat org.apache.spark.scheduler.DAGScheduler.abortStage(DAGScheduler.scala:2007)\n\tat org.apache.spark.scheduler.DAGScheduler.$anonfun$handleTaskSetFailed$1(DAGScheduler.scala:973)\n\tat org.apache.spark.scheduler.DAGScheduler.$anonfun$handleTaskSetFailed$1$adapted(DAGScheduler.scala:973)\n\tat scala.Option.foreach(Option.scala:407)\n\tat org.apache.spark.scheduler.DAGScheduler.handleTaskSetFailed(DAGScheduler.scala:973)\n\tat org.apache.spark.scheduler.DAGSchedulerEventProcessLoop.doOnReceive(DAGScheduler.scala:2239)\n\tat org.apache.spark.scheduler.DAGSchedulerEventProcessLoop.onReceive(DAGScheduler.scala:2188)\n\tat org.apache.spark.scheduler.DAGSchedulerEventProcessLoop.onReceive(DAGScheduler.scala:2177)\n\tat org.apache.spark.util.EventLoop$$anon$1.run(EventLoop.scala:49)\n\tat org.apache.spark.scheduler.DAGScheduler.runJob(DAGScheduler.scala:775)\n\tat org.apache.spark.SparkContext.runJob(SparkContext.scala:2099)\n\tat org.apache.spark.SparkContext.runJob(SparkContext.scala:2120)\n\tat org.apache.spark.SparkContext.runJob(SparkContext.scala:2139)\n\tat org.apache.spark.api.python.PythonRDD$.runJob(PythonRDD.scala:154)\n\tat org.apache.spark.api.python.PythonRDD.runJob(PythonRDD.scala)\n\tat java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)\n\tat java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)\n\tat java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)\n\tat java.base/java.lang.reflect.Method.invoke(Method.java:566)\n\tat py4j.reflection.MethodInvoker.invoke(MethodInvoker.java:244)\n\tat py4j.reflection.ReflectionEngine.invoke(ReflectionEngine.java:357)\n\tat py4j.Gateway.invoke(Gateway.java:282)\n\tat py4j.commands.AbstractCommand.invokeMethod(AbstractCommand.java:132)\n\tat py4j.commands.CallCommand.execute(CallCommand.java:79)\n\tat py4j.GatewayConnection.run(GatewayConnection.java:238)\n\tat java.base/java.lang.Thread.run(Thread.java:834)\nCaused by: org.apache.spark.api.python.PythonException: Traceback (most recent call last):\n  File \"/usr/local/Cellar/apache-spark/3.0.1/libexec/python/lib/pyspark.zip/pyspark/worker.py\", line 605, in main\n    process()\n  File \"/usr/local/Cellar/apache-spark/3.0.1/libexec/python/lib/pyspark.zip/pyspark/worker.py\", line 597, in process\n    serializer.dump_stream(out_iter, outfile)\n  File \"/usr/local/Cellar/apache-spark/3.0.1/libexec/python/lib/pyspark.zip/pyspark/serializers.py\", line 271, in dump_stream\n    vs = list(itertools.islice(iterator, batch))\n  File \"/usr/local/Cellar/apache-spark/3.0.1/libexec/python/pyspark/rdd.py\", line 1440, in takeUpToNumLeft\n    yield next(iterator)\n  File \"/usr/local/Cellar/apache-spark/3.0.1/libexec/python/lib/pyspark.zip/pyspark/util.py\", line 107, in wrapper\n    return f(*args, **kwargs)\n  File \"<ipython-input-89-913fc787ca91>\", line 15, in <lambda>\n  File \"<ipython-input-89-913fc787ca91>\", line 7, in parser_xml\n  File \"/Users/madalina-alinaracovita/opt/anaconda3/lib/python3.8/xml/etree/ElementTree.py\", line 1320, in XML\n    parser.feed(text)\n  File \"<string>\", line None\nxml.etree.ElementTree.ParseError: not well-formed (invalid token): line 107, column 29\n\n\tat org.apache.spark.api.python.BasePythonRunner$ReaderIterator.handlePythonException(PythonRunner.scala:503)\n\tat org.apache.spark.api.python.PythonRunner$$anon$3.read(PythonRunner.scala:638)\n\tat org.apache.spark.api.python.PythonRunner$$anon$3.read(PythonRunner.scala:621)\n\tat org.apache.spark.api.python.BasePythonRunner$ReaderIterator.hasNext(PythonRunner.scala:456)\n\tat org.apache.spark.InterruptibleIterator.hasNext(InterruptibleIterator.scala:37)\n\tat scala.collection.Iterator.foreach(Iterator.scala:941)\n\tat scala.collection.Iterator.foreach$(Iterator.scala:941)\n\tat org.apache.spark.InterruptibleIterator.foreach(InterruptibleIterator.scala:28)\n\tat scala.collection.generic.Growable.$plus$plus$eq(Growable.scala:62)\n\tat scala.collection.generic.Growable.$plus$plus$eq$(Growable.scala:53)\n\tat scala.collection.mutable.ArrayBuffer.$plus$plus$eq(ArrayBuffer.scala:105)\n\tat scala.collection.mutable.ArrayBuffer.$plus$plus$eq(ArrayBuffer.scala:49)\n\tat scala.collection.TraversableOnce.to(TraversableOnce.scala:315)\n\tat scala.collection.TraversableOnce.to$(TraversableOnce.scala:313)\n\tat org.apache.spark.InterruptibleIterator.to(InterruptibleIterator.scala:28)\n\tat scala.collection.TraversableOnce.toBuffer(TraversableOnce.scala:307)\n\tat scala.collection.TraversableOnce.toBuffer$(TraversableOnce.scala:307)\n\tat org.apache.spark.InterruptibleIterator.toBuffer(InterruptibleIterator.scala:28)\n\tat scala.collection.TraversableOnce.toArray(TraversableOnce.scala:294)\n\tat scala.collection.TraversableOnce.toArray$(TraversableOnce.scala:288)\n\tat org.apache.spark.InterruptibleIterator.toArray(InterruptibleIterator.scala:28)\n\tat org.apache.spark.api.python.PythonRDD$.$anonfun$runJob$1(PythonRDD.scala:154)\n\tat org.apache.spark.SparkContext.$anonfun$runJob$5(SparkContext.scala:2139)\n\tat org.apache.spark.scheduler.ResultTask.runTask(ResultTask.scala:90)\n\tat org.apache.spark.scheduler.Task.run(Task.scala:127)\n\tat org.apache.spark.executor.Executor$TaskRunner.$anonfun$run$3(Executor.scala:446)\n\tat org.apache.spark.util.Utils$.tryWithSafeFinally(Utils.scala:1377)\n\tat org.apache.spark.executor.Executor$TaskRunner.run(Executor.scala:449)\n\tat java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1128)\n\tat java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:628)\n\t... 1 more\n",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mPy4JJavaError\u001b[0m                             Traceback (most recent call last)",
      "\u001b[0;32m<ipython-input-89-913fc787ca91>\u001b[0m in \u001b[0;36m<module>\u001b[0;34m\u001b[0m\n\u001b[1;32m     14\u001b[0m \u001b[0mrdd\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0msc\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mwholeTextFiles\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mfile_path\u001b[0m \u001b[0;34m+\u001b[0m \u001b[0;34m\"/*.xml.gz\"\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m     15\u001b[0m \u001b[0mdata\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mrdd\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mmap\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;32mlambda\u001b[0m \u001b[0mstring_file\u001b[0m\u001b[0;34m:\u001b[0m \u001b[0mparser_xml\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mstring_file\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0;36m1\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m---> 16\u001b[0;31m \u001b[0mmy_df\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mspark\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mcreateDataFrame\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mdata\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mschema\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0;32mNone\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0msamplingRatio\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0;32mNone\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mverifySchema\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0;32mTrue\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m     17\u001b[0m \u001b[0mmy_df\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mtoPandas\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m/usr/local/Cellar/apache-spark/3.0.1/libexec/python/pyspark/sql/session.py\u001b[0m in \u001b[0;36mcreateDataFrame\u001b[0;34m(self, data, schema, samplingRatio, verifySchema)\u001b[0m\n\u001b[1;32m    603\u001b[0m             return super(SparkSession, self).createDataFrame(\n\u001b[1;32m    604\u001b[0m                 data, schema, samplingRatio, verifySchema)\n\u001b[0;32m--> 605\u001b[0;31m         \u001b[0;32mreturn\u001b[0m \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m_create_dataframe\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mdata\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mschema\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0msamplingRatio\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mverifySchema\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m    606\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    607\u001b[0m     \u001b[0;32mdef\u001b[0m \u001b[0m_create_dataframe\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mself\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mdata\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mschema\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0msamplingRatio\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mverifySchema\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m/usr/local/Cellar/apache-spark/3.0.1/libexec/python/pyspark/sql/session.py\u001b[0m in \u001b[0;36m_create_dataframe\u001b[0;34m(self, data, schema, samplingRatio, verifySchema)\u001b[0m\n\u001b[1;32m    626\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    627\u001b[0m         \u001b[0;32mif\u001b[0m \u001b[0misinstance\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mdata\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mRDD\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m--> 628\u001b[0;31m             \u001b[0mrdd\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mschema\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m_createFromRDD\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mdata\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mmap\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mprepare\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mschema\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0msamplingRatio\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m    629\u001b[0m         \u001b[0;32melse\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    630\u001b[0m             \u001b[0mrdd\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mschema\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m_createFromLocal\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mmap\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mprepare\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mdata\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mschema\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m/usr/local/Cellar/apache-spark/3.0.1/libexec/python/pyspark/sql/session.py\u001b[0m in \u001b[0;36m_createFromRDD\u001b[0;34m(self, rdd, schema, samplingRatio)\u001b[0m\n\u001b[1;32m    423\u001b[0m         \"\"\"\n\u001b[1;32m    424\u001b[0m         \u001b[0;32mif\u001b[0m \u001b[0mschema\u001b[0m \u001b[0;32mis\u001b[0m \u001b[0;32mNone\u001b[0m \u001b[0;32mor\u001b[0m \u001b[0misinstance\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mschema\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m(\u001b[0m\u001b[0mlist\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mtuple\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m--> 425\u001b[0;31m             \u001b[0mstruct\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m_inferSchema\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mrdd\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0msamplingRatio\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mnames\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0mschema\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m    426\u001b[0m             \u001b[0mconverter\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0m_create_converter\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mstruct\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    427\u001b[0m             \u001b[0mrdd\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mrdd\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mmap\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mconverter\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m/usr/local/Cellar/apache-spark/3.0.1/libexec/python/pyspark/sql/session.py\u001b[0m in \u001b[0;36m_inferSchema\u001b[0;34m(self, rdd, samplingRatio, names)\u001b[0m\n\u001b[1;32m    394\u001b[0m         \u001b[0;34m:\u001b[0m\u001b[0;32mreturn\u001b[0m\u001b[0;34m:\u001b[0m \u001b[0;34m:\u001b[0m\u001b[0;32mclass\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;31m`\u001b[0m\u001b[0mpyspark\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0msql\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mtypes\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mStructType\u001b[0m\u001b[0;31m`\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    395\u001b[0m         \"\"\"\n\u001b[0;32m--> 396\u001b[0;31m         \u001b[0mfirst\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mrdd\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mfirst\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m    397\u001b[0m         \u001b[0;32mif\u001b[0m \u001b[0;32mnot\u001b[0m \u001b[0mfirst\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    398\u001b[0m             raise ValueError(\"The first row in RDD is empty, \"\n",
      "\u001b[0;32m/usr/local/Cellar/apache-spark/3.0.1/libexec/python/pyspark/rdd.py\u001b[0m in \u001b[0;36mfirst\u001b[0;34m(self)\u001b[0m\n\u001b[1;32m   1462\u001b[0m         \u001b[0mValueError\u001b[0m\u001b[0;34m:\u001b[0m \u001b[0mRDD\u001b[0m \u001b[0;32mis\u001b[0m \u001b[0mempty\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   1463\u001b[0m         \"\"\"\n\u001b[0;32m-> 1464\u001b[0;31m         \u001b[0mrs\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mtake\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;36m1\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m   1465\u001b[0m         \u001b[0;32mif\u001b[0m \u001b[0mrs\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   1466\u001b[0m             \u001b[0;32mreturn\u001b[0m \u001b[0mrs\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0;36m0\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m/usr/local/Cellar/apache-spark/3.0.1/libexec/python/pyspark/rdd.py\u001b[0m in \u001b[0;36mtake\u001b[0;34m(self, num)\u001b[0m\n\u001b[1;32m   1444\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   1445\u001b[0m             \u001b[0mp\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mrange\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mpartsScanned\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mmin\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mpartsScanned\u001b[0m \u001b[0;34m+\u001b[0m \u001b[0mnumPartsToTry\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mtotalParts\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m-> 1446\u001b[0;31m             \u001b[0mres\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mcontext\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mrunJob\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mself\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mtakeUpToNumLeft\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mp\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m   1447\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   1448\u001b[0m             \u001b[0mitems\u001b[0m \u001b[0;34m+=\u001b[0m \u001b[0mres\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m/usr/local/Cellar/apache-spark/3.0.1/libexec/python/pyspark/context.py\u001b[0m in \u001b[0;36mrunJob\u001b[0;34m(self, rdd, partitionFunc, partitions, allowLocal)\u001b[0m\n\u001b[1;32m   1116\u001b[0m         \u001b[0;31m# SparkContext#runJob.\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   1117\u001b[0m         \u001b[0mmappedRDD\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mrdd\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mmapPartitions\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mpartitionFunc\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m-> 1118\u001b[0;31m         \u001b[0msock_info\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m_jvm\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mPythonRDD\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mrunJob\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m_jsc\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0msc\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mmappedRDD\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m_jrdd\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mpartitions\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m   1119\u001b[0m         \u001b[0;32mreturn\u001b[0m \u001b[0mlist\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0m_load_from_socket\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0msock_info\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mmappedRDD\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m_jrdd_deserializer\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   1120\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m/usr/local/Cellar/apache-spark/3.0.1/libexec/python/lib/py4j-0.10.9-src.zip/py4j/java_gateway.py\u001b[0m in \u001b[0;36m__call__\u001b[0;34m(self, *args)\u001b[0m\n\u001b[1;32m   1302\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   1303\u001b[0m         \u001b[0manswer\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mgateway_client\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0msend_command\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mcommand\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m-> 1304\u001b[0;31m         return_value = get_return_value(\n\u001b[0m\u001b[1;32m   1305\u001b[0m             answer, self.gateway_client, self.target_id, self.name)\n\u001b[1;32m   1306\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m/usr/local/Cellar/apache-spark/3.0.1/libexec/python/pyspark/sql/utils.py\u001b[0m in \u001b[0;36mdeco\u001b[0;34m(*a, **kw)\u001b[0m\n\u001b[1;32m    126\u001b[0m     \u001b[0;32mdef\u001b[0m \u001b[0mdeco\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m*\u001b[0m\u001b[0ma\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m**\u001b[0m\u001b[0mkw\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    127\u001b[0m         \u001b[0;32mtry\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m--> 128\u001b[0;31m             \u001b[0;32mreturn\u001b[0m \u001b[0mf\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m*\u001b[0m\u001b[0ma\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m**\u001b[0m\u001b[0mkw\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m    129\u001b[0m         \u001b[0;32mexcept\u001b[0m \u001b[0mpy4j\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mprotocol\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mPy4JJavaError\u001b[0m \u001b[0;32mas\u001b[0m \u001b[0me\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    130\u001b[0m             \u001b[0mconverted\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mconvert_exception\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0me\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mjava_exception\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m/usr/local/Cellar/apache-spark/3.0.1/libexec/python/lib/py4j-0.10.9-src.zip/py4j/protocol.py\u001b[0m in \u001b[0;36mget_return_value\u001b[0;34m(answer, gateway_client, target_id, name)\u001b[0m\n\u001b[1;32m    324\u001b[0m             \u001b[0mvalue\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mOUTPUT_CONVERTER\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0mtype\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0manswer\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0;36m2\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mgateway_client\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    325\u001b[0m             \u001b[0;32mif\u001b[0m \u001b[0manswer\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0;36m1\u001b[0m\u001b[0;34m]\u001b[0m \u001b[0;34m==\u001b[0m \u001b[0mREFERENCE_TYPE\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m--> 326\u001b[0;31m                 raise Py4JJavaError(\n\u001b[0m\u001b[1;32m    327\u001b[0m                     \u001b[0;34m\"An error occurred while calling {0}{1}{2}.\\n\"\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    328\u001b[0m                     format(target_id, \".\", name), value)\n",
      "\u001b[0;31mPy4JJavaError\u001b[0m: An error occurred while calling z:org.apache.spark.api.python.PythonRDD.runJob.\n: org.apache.spark.SparkException: Job aborted due to stage failure: Task 0 in stage 94.0 failed 1 times, most recent failure: Lost task 0.0 in stage 94.0 (TID 458, madalinlinasmbp, executor driver): org.apache.spark.api.python.PythonException: Traceback (most recent call last):\n  File \"/usr/local/Cellar/apache-spark/3.0.1/libexec/python/lib/pyspark.zip/pyspark/worker.py\", line 605, in main\n    process()\n  File \"/usr/local/Cellar/apache-spark/3.0.1/libexec/python/lib/pyspark.zip/pyspark/worker.py\", line 597, in process\n    serializer.dump_stream(out_iter, outfile)\n  File \"/usr/local/Cellar/apache-spark/3.0.1/libexec/python/lib/pyspark.zip/pyspark/serializers.py\", line 271, in dump_stream\n    vs = list(itertools.islice(iterator, batch))\n  File \"/usr/local/Cellar/apache-spark/3.0.1/libexec/python/pyspark/rdd.py\", line 1440, in takeUpToNumLeft\n    yield next(iterator)\n  File \"/usr/local/Cellar/apache-spark/3.0.1/libexec/python/lib/pyspark.zip/pyspark/util.py\", line 107, in wrapper\n    return f(*args, **kwargs)\n  File \"<ipython-input-89-913fc787ca91>\", line 15, in <lambda>\n  File \"<ipython-input-89-913fc787ca91>\", line 7, in parser_xml\n  File \"/Users/madalina-alinaracovita/opt/anaconda3/lib/python3.8/xml/etree/ElementTree.py\", line 1320, in XML\n    parser.feed(text)\n  File \"<string>\", line None\nxml.etree.ElementTree.ParseError: not well-formed (invalid token): line 107, column 29\n\n\tat org.apache.spark.api.python.BasePythonRunner$ReaderIterator.handlePythonException(PythonRunner.scala:503)\n\tat org.apache.spark.api.python.PythonRunner$$anon$3.read(PythonRunner.scala:638)\n\tat org.apache.spark.api.python.PythonRunner$$anon$3.read(PythonRunner.scala:621)\n\tat org.apache.spark.api.python.BasePythonRunner$ReaderIterator.hasNext(PythonRunner.scala:456)\n\tat org.apache.spark.InterruptibleIterator.hasNext(InterruptibleIterator.scala:37)\n\tat scala.collection.Iterator.foreach(Iterator.scala:941)\n\tat scala.collection.Iterator.foreach$(Iterator.scala:941)\n\tat org.apache.spark.InterruptibleIterator.foreach(InterruptibleIterator.scala:28)\n\tat scala.collection.generic.Growable.$plus$plus$eq(Growable.scala:62)\n\tat scala.collection.generic.Growable.$plus$plus$eq$(Growable.scala:53)\n\tat scala.collection.mutable.ArrayBuffer.$plus$plus$eq(ArrayBuffer.scala:105)\n\tat scala.collection.mutable.ArrayBuffer.$plus$plus$eq(ArrayBuffer.scala:49)\n\tat scala.collection.TraversableOnce.to(TraversableOnce.scala:315)\n\tat scala.collection.TraversableOnce.to$(TraversableOnce.scala:313)\n\tat org.apache.spark.InterruptibleIterator.to(InterruptibleIterator.scala:28)\n\tat scala.collection.TraversableOnce.toBuffer(TraversableOnce.scala:307)\n\tat scala.collection.TraversableOnce.toBuffer$(TraversableOnce.scala:307)\n\tat org.apache.spark.InterruptibleIterator.toBuffer(InterruptibleIterator.scala:28)\n\tat scala.collection.TraversableOnce.toArray(TraversableOnce.scala:294)\n\tat scala.collection.TraversableOnce.toArray$(TraversableOnce.scala:288)\n\tat org.apache.spark.InterruptibleIterator.toArray(InterruptibleIterator.scala:28)\n\tat org.apache.spark.api.python.PythonRDD$.$anonfun$runJob$1(PythonRDD.scala:154)\n\tat org.apache.spark.SparkContext.$anonfun$runJob$5(SparkContext.scala:2139)\n\tat org.apache.spark.scheduler.ResultTask.runTask(ResultTask.scala:90)\n\tat org.apache.spark.scheduler.Task.run(Task.scala:127)\n\tat org.apache.spark.executor.Executor$TaskRunner.$anonfun$run$3(Executor.scala:446)\n\tat org.apache.spark.util.Utils$.tryWithSafeFinally(Utils.scala:1377)\n\tat org.apache.spark.executor.Executor$TaskRunner.run(Executor.scala:449)\n\tat java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1128)\n\tat java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:628)\n\tat java.base/java.lang.Thread.run(Thread.java:834)\n\nDriver stacktrace:\n\tat org.apache.spark.scheduler.DAGScheduler.failJobAndIndependentStages(DAGScheduler.scala:2059)\n\tat org.apache.spark.scheduler.DAGScheduler.$anonfun$abortStage$2(DAGScheduler.scala:2008)\n\tat org.apache.spark.scheduler.DAGScheduler.$anonfun$abortStage$2$adapted(DAGScheduler.scala:2007)\n\tat scala.collection.mutable.ResizableArray.foreach(ResizableArray.scala:62)\n\tat scala.collection.mutable.ResizableArray.foreach$(ResizableArray.scala:55)\n\tat scala.collection.mutable.ArrayBuffer.foreach(ArrayBuffer.scala:49)\n\tat org.apache.spark.scheduler.DAGScheduler.abortStage(DAGScheduler.scala:2007)\n\tat org.apache.spark.scheduler.DAGScheduler.$anonfun$handleTaskSetFailed$1(DAGScheduler.scala:973)\n\tat org.apache.spark.scheduler.DAGScheduler.$anonfun$handleTaskSetFailed$1$adapted(DAGScheduler.scala:973)\n\tat scala.Option.foreach(Option.scala:407)\n\tat org.apache.spark.scheduler.DAGScheduler.handleTaskSetFailed(DAGScheduler.scala:973)\n\tat org.apache.spark.scheduler.DAGSchedulerEventProcessLoop.doOnReceive(DAGScheduler.scala:2239)\n\tat org.apache.spark.scheduler.DAGSchedulerEventProcessLoop.onReceive(DAGScheduler.scala:2188)\n\tat org.apache.spark.scheduler.DAGSchedulerEventProcessLoop.onReceive(DAGScheduler.scala:2177)\n\tat org.apache.spark.util.EventLoop$$anon$1.run(EventLoop.scala:49)\n\tat org.apache.spark.scheduler.DAGScheduler.runJob(DAGScheduler.scala:775)\n\tat org.apache.spark.SparkContext.runJob(SparkContext.scala:2099)\n\tat org.apache.spark.SparkContext.runJob(SparkContext.scala:2120)\n\tat org.apache.spark.SparkContext.runJob(SparkContext.scala:2139)\n\tat org.apache.spark.api.python.PythonRDD$.runJob(PythonRDD.scala:154)\n\tat org.apache.spark.api.python.PythonRDD.runJob(PythonRDD.scala)\n\tat java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)\n\tat java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)\n\tat java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)\n\tat java.base/java.lang.reflect.Method.invoke(Method.java:566)\n\tat py4j.reflection.MethodInvoker.invoke(MethodInvoker.java:244)\n\tat py4j.reflection.ReflectionEngine.invoke(ReflectionEngine.java:357)\n\tat py4j.Gateway.invoke(Gateway.java:282)\n\tat py4j.commands.AbstractCommand.invokeMethod(AbstractCommand.java:132)\n\tat py4j.commands.CallCommand.execute(CallCommand.java:79)\n\tat py4j.GatewayConnection.run(GatewayConnection.java:238)\n\tat java.base/java.lang.Thread.run(Thread.java:834)\nCaused by: org.apache.spark.api.python.PythonException: Traceback (most recent call last):\n  File \"/usr/local/Cellar/apache-spark/3.0.1/libexec/python/lib/pyspark.zip/pyspark/worker.py\", line 605, in main\n    process()\n  File \"/usr/local/Cellar/apache-spark/3.0.1/libexec/python/lib/pyspark.zip/pyspark/worker.py\", line 597, in process\n    serializer.dump_stream(out_iter, outfile)\n  File \"/usr/local/Cellar/apache-spark/3.0.1/libexec/python/lib/pyspark.zip/pyspark/serializers.py\", line 271, in dump_stream\n    vs = list(itertools.islice(iterator, batch))\n  File \"/usr/local/Cellar/apache-spark/3.0.1/libexec/python/pyspark/rdd.py\", line 1440, in takeUpToNumLeft\n    yield next(iterator)\n  File \"/usr/local/Cellar/apache-spark/3.0.1/libexec/python/lib/pyspark.zip/pyspark/util.py\", line 107, in wrapper\n    return f(*args, **kwargs)\n  File \"<ipython-input-89-913fc787ca91>\", line 15, in <lambda>\n  File \"<ipython-input-89-913fc787ca91>\", line 7, in parser_xml\n  File \"/Users/madalina-alinaracovita/opt/anaconda3/lib/python3.8/xml/etree/ElementTree.py\", line 1320, in XML\n    parser.feed(text)\n  File \"<string>\", line None\nxml.etree.ElementTree.ParseError: not well-formed (invalid token): line 107, column 29\n\n\tat org.apache.spark.api.python.BasePythonRunner$ReaderIterator.handlePythonException(PythonRunner.scala:503)\n\tat org.apache.spark.api.python.PythonRunner$$anon$3.read(PythonRunner.scala:638)\n\tat org.apache.spark.api.python.PythonRunner$$anon$3.read(PythonRunner.scala:621)\n\tat org.apache.spark.api.python.BasePythonRunner$ReaderIterator.hasNext(PythonRunner.scala:456)\n\tat org.apache.spark.InterruptibleIterator.hasNext(InterruptibleIterator.scala:37)\n\tat scala.collection.Iterator.foreach(Iterator.scala:941)\n\tat scala.collection.Iterator.foreach$(Iterator.scala:941)\n\tat org.apache.spark.InterruptibleIterator.foreach(InterruptibleIterator.scala:28)\n\tat scala.collection.generic.Growable.$plus$plus$eq(Growable.scala:62)\n\tat scala.collection.generic.Growable.$plus$plus$eq$(Growable.scala:53)\n\tat scala.collection.mutable.ArrayBuffer.$plus$plus$eq(ArrayBuffer.scala:105)\n\tat scala.collection.mutable.ArrayBuffer.$plus$plus$eq(ArrayBuffer.scala:49)\n\tat scala.collection.TraversableOnce.to(TraversableOnce.scala:315)\n\tat scala.collection.TraversableOnce.to$(TraversableOnce.scala:313)\n\tat org.apache.spark.InterruptibleIterator.to(InterruptibleIterator.scala:28)\n\tat scala.collection.TraversableOnce.toBuffer(TraversableOnce.scala:307)\n\tat scala.collection.TraversableOnce.toBuffer$(TraversableOnce.scala:307)\n\tat org.apache.spark.InterruptibleIterator.toBuffer(InterruptibleIterator.scala:28)\n\tat scala.collection.TraversableOnce.toArray(TraversableOnce.scala:294)\n\tat scala.collection.TraversableOnce.toArray$(TraversableOnce.scala:288)\n\tat org.apache.spark.InterruptibleIterator.toArray(InterruptibleIterator.scala:28)\n\tat org.apache.spark.api.python.PythonRDD$.$anonfun$runJob$1(PythonRDD.scala:154)\n\tat org.apache.spark.SparkContext.$anonfun$runJob$5(SparkContext.scala:2139)\n\tat org.apache.spark.scheduler.ResultTask.runTask(ResultTask.scala:90)\n\tat org.apache.spark.scheduler.Task.run(Task.scala:127)\n\tat org.apache.spark.executor.Executor$TaskRunner.$anonfun$run$3(Executor.scala:446)\n\tat org.apache.spark.util.Utils$.tryWithSafeFinally(Utils.scala:1377)\n\tat org.apache.spark.executor.Executor$TaskRunner.run(Executor.scala:449)\n\tat java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1128)\n\tat java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:628)\n\t... 1 more\n"
     ]
    }
   ],
   "source": [
    "df.withColumn(\"LineItem\", explode($\"RetailTransaction.LineItem\"))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>counts</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>98</th>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>99</th>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>100</th>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>101</th>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>102</th>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>103 rows × 1 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "    counts\n",
       "0     None\n",
       "1     None\n",
       "2     None\n",
       "3     None\n",
       "4     None\n",
       "..     ...\n",
       "98    None\n",
       "99    None\n",
       "100   None\n",
       "101   None\n",
       "102   None\n",
       "\n",
       "[103 rows x 1 columns]"
      ]
     },
     "execution_count": 20,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "from pyspark.sql.types import *\n",
    "\n",
    "customSchema = StructType([\n",
    "    StructField(\"counts\", StringType(), True)])\n",
    "\n",
    "df = spark.read \\\n",
    "    .format('xml') \\\n",
    "    .options(rowTag='record').options(rootTag)\\\n",
    "    .load(file_path, schema = customSchema)\n",
    "\n",
    "df.select(\"counts\").toPandas()"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.3"
  },
  "toc": {
   "base_numbering": 1,
   "nav_menu": {},
   "number_sections": false,
   "sideBar": true,
   "skip_h1_title": false,
   "title_cell": "Data parsing",
   "title_sidebar": "Contents",
   "toc_cell": true,
   "toc_position": {},
   "toc_section_display": true,
   "toc_window_display": true
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
